diff --git a/xbmc/GUILargeTextureManager.cpp b/xbmc/GUILargeTextureManager.cpp
index 80b762b20..6a3bfc6a8 100644
--- a/xbmc/GUILargeTextureManager.cpp
+++ b/xbmc/GUILargeTextureManager.cpp
@@ -13,6 +13,8 @@
 #include "commons/ilog.h"
 #include "guilib/GUIComponent.h"
 #include "guilib/Texture.h"
+#include "settings/AdvancedSettings.h"
+#include "settings/SettingsComponent.h"
 #include "utils/JobManager.h"
 #include "utils/TimeUtils.h"
 #include "utils/log.h"
@@ -65,6 +67,9 @@ bool CImageLoader::DoWork()
       if (needsChecking)
         CServiceBroker::GetTextureCache()->BackgroundCacheImage(texturePath);
 
+      if (CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_guiAsyncTextureUpload)
+        m_texture->LoadToGPUAsync();
+
       return true;
     }
 
@@ -77,7 +82,14 @@ bool CImageLoader::DoWork()
 
   // not in our texture cache or it failed to load from it, so try and load directly and then cache the result
   CServiceBroker::GetTextureCache()->CacheImage(texturePath, &m_texture);
-  return (m_texture != NULL);
+
+  if (!m_texture)
+    return false;
+
+  if (CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_guiAsyncTextureUpload)
+    m_texture->LoadToGPUAsync();
+
+  return true;
 }
 
 CGUILargeTextureManager::CLargeTexture::CLargeTexture(const std::string &path):
diff --git a/xbmc/guilib/CMakeLists.txt b/xbmc/guilib/CMakeLists.txt
index 3e936cd9b..a4c3957f5 100644
--- a/xbmc/guilib/CMakeLists.txt
+++ b/xbmc/guilib/CMakeLists.txt
@@ -66,6 +66,7 @@ set(SOURCES DDSImage.cpp
             TextureBundle.cpp
             TextureBundleXBT.cpp
             Texture.cpp
+            TextureBase.cpp
             TextureManager.cpp
             VisibleEffect.cpp
             XBTF.cpp
@@ -148,6 +149,7 @@ set(HEADERS DDSImage.h
             LocalizeStrings.h
             StereoscopicsManager.h
             Texture.h
+            TextureBase.h
             TextureBundle.h
             TextureBundleXBT.h
             TextureManager.h
diff --git a/xbmc/guilib/GUIBorderedImage.cpp b/xbmc/guilib/GUIBorderedImage.cpp
index 7c48ddb9e..b3290f8cf 100644
--- a/xbmc/guilib/GUIBorderedImage.cpp
+++ b/xbmc/guilib/GUIBorderedImage.cpp
@@ -39,12 +39,12 @@ CGUIBorderedImage::CGUIBorderedImage(const CGUIBorderedImage& right)
 void CGUIBorderedImage::Process(unsigned int currentTime, CDirtyRegionList &dirtyregions)
 {
   CGUIImage::Process(currentTime, dirtyregions);
-  if (!m_borderImage->GetFileName().empty() && m_texture->ReadyToRender())
+  if (!m_borderImage->GetFileName().empty() && m_textureCurrent->ReadyToRender())
   {
-    CRect rect = CRect(m_texture->GetXPosition(), m_texture->GetYPosition(),
-                       m_texture->GetXPosition() + m_texture->GetWidth(),
-                       m_texture->GetYPosition() + m_texture->GetHeight());
-    rect.Intersect(m_texture->GetRenderRect());
+    CRect rect = CRect(m_textureCurrent->GetXPosition(), m_textureCurrent->GetYPosition(),
+                       m_textureCurrent->GetXPosition() + m_textureCurrent->GetWidth(),
+                       m_textureCurrent->GetYPosition() + m_textureCurrent->GetHeight());
+    rect.Intersect(m_textureCurrent->GetRenderRect());
     m_borderImage->SetPosition(rect.x1 - m_borderSize.x1, rect.y1 - m_borderSize.y1);
     m_borderImage->SetWidth(rect.Width() + m_borderSize.x1 + m_borderSize.x2);
     m_borderImage->SetHeight(rect.Height() + m_borderSize.y1 + m_borderSize.y2);
@@ -56,9 +56,17 @@ void CGUIBorderedImage::Process(unsigned int currentTime, CDirtyRegionList &dirt
 
 void CGUIBorderedImage::Render()
 {
-  if (!m_borderImage->GetFileName().empty() && m_texture->ReadyToRender())
-    m_borderImage->Render();
-  CGUIImage::Render();
+  bool renderFrontToBack = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderOrder() ==
+                           RENDER_ORDER_FRONT_TO_BACK;
+
+  if (renderFrontToBack)
+    CGUIImage::Render();
+
+  if (!m_borderImage->GetFileName().empty() && m_textureCurrent->ReadyToRender())
+    m_borderImage->Render(-1);
+
+  if (!renderFrontToBack)
+    CGUIImage::Render();
 }
 
 CRect CGUIBorderedImage::CalcRenderRegion() const
diff --git a/xbmc/guilib/GUIColorButtonControl.cpp b/xbmc/guilib/GUIColorButtonControl.cpp
index 1e40667cf..4917af470 100644
--- a/xbmc/guilib/GUIColorButtonControl.cpp
+++ b/xbmc/guilib/GUIColorButtonControl.cpp
@@ -57,7 +57,6 @@ CGUIColorButtonControl::CGUIColorButtonControl(const CGUIColorButtonControl& con
 void CGUIColorButtonControl::Render()
 {
   CGUIButtonControl::Render();
-  RenderInfoText();
   if (IsDisabled())
     m_imgColorDisabledMask->Render();
   else
@@ -180,9 +179,13 @@ bool CGUIColorButtonControl::UpdateColors(const CGUIListItem* item)
   return changed;
 }
 
-void CGUIColorButtonControl::RenderInfoText()
+void CGUIColorButtonControl::RenderText()
 {
-  m_labelInfo.Render();
+  if (CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderOrder() !=
+      RENDER_ORDER_FRONT_TO_BACK)
+    m_labelInfo.Render();
+
+  CGUIButtonControl::RenderText();
 }
 
 void CGUIColorButtonControl::ProcessInfoText(unsigned int currentTime)
diff --git a/xbmc/guilib/GUIColorButtonControl.h b/xbmc/guilib/GUIColorButtonControl.h
index 828898cc1..d8ebdd042 100644
--- a/xbmc/guilib/GUIColorButtonControl.h
+++ b/xbmc/guilib/GUIColorButtonControl.h
@@ -60,7 +60,7 @@ public:
 protected:
   bool UpdateColors(const CGUIListItem* item) override;
   void ProcessInfoText(unsigned int currentTime);
-  void RenderInfoText();
+  void RenderText() override;
   CGUILabel::COLOR GetTextColor() const override;
   std::unique_ptr<CGUITexture> m_imgColorMask;
   std::unique_ptr<CGUITexture> m_imgColorDisabledMask;
diff --git a/xbmc/guilib/GUIFontCache.cpp b/xbmc/guilib/GUIFontCache.cpp
index 70f17acc3..d92643676 100644
--- a/xbmc/guilib/GUIFontCache.cpp
+++ b/xbmc/guilib/GUIFontCache.cpp
@@ -65,8 +65,10 @@ class CGUIFontCacheImpl
       {
         if (ageit->second == it)
         {
-          ageMap.erase(ageit);
-          ageMap.insert(typename AgeMap::value_type(now, it));
+          auto node = ageMap.extract(ageit);
+          node.key() = now;
+          node.mapped() = it;
+          ageMap.insert(std::move(node));
           it->second->m_lastUsed = now;
           return;
         }
diff --git a/xbmc/guilib/GUIImage.cpp b/xbmc/guilib/GUIImage.cpp
index 9b7ad23de..113297792 100644
--- a/xbmc/guilib/GUIImage.cpp
+++ b/xbmc/guilib/GUIImage.cpp
@@ -23,22 +23,23 @@ CGUIImage::CGUIImage(int parentID,
                      float height,
                      const CTextureInfo& texture)
   : CGUIControl(parentID, controlID, posX, posY, width, height),
-    m_texture(CGUITexture::CreateTexture(posX, posY, width, height, texture))
+    m_textureCurrent(CGUITexture::CreateTexture(posX, posY, width, height, texture)),
+    m_textureNext(CGUITexture::CreateTexture(posX, posY, width, height, texture))
 {
   m_crossFadeTime = 0;
   m_currentFadeTime = 0;
   m_lastRenderTime = 0;
   ControlType = GUICONTROL_IMAGE;
   m_bDynamicResourceAlloc=false;
+  m_textureNext->SetFileName("");
 }
 
 CGUIImage::CGUIImage(const CGUIImage& left)
   : CGUIControl(left),
     m_image(left.m_image),
     m_info(left.m_info),
-    m_texture(left.m_texture->Clone()),
-    m_fadingTextures(),
-    m_currentTexture(),
+    m_textureCurrent(left.m_textureCurrent->Clone()),
+    m_textureNext(left.m_textureNext->Clone()),
     m_currentFallback()
 {
   m_crossFadeTime = left.m_crossFadeTime;
@@ -62,16 +63,15 @@ void CGUIImage::UpdateVisibility(const CGUIListItem *item)
 
 void CGUIImage::UpdateDiffuseColor(const CGUIListItem* item)
 {
-  if (m_texture->SetDiffuseColor(m_diffuseColor, item))
-  {
+  if (m_textureCurrent->SetDiffuseColor(m_diffuseColor, item))
     MarkDirtyRegion();
-  }
+  m_textureNext->SetDiffuseColor(m_diffuseColor, item);
 }
 
 void CGUIImage::UpdateInfo(const CGUIListItem *item)
 {
   // The texture may also depend on info conditions. Update the diffuse color in that case.
-  if (m_texture->GetDiffuseColor().HasInfo())
+  if (m_textureCurrent->GetDiffuseColor().HasInfo())
     UpdateDiffuseColor(item);
 
   if (m_info.IsConstant())
@@ -90,125 +90,204 @@ void CGUIImage::UpdateInfo(const CGUIListItem *item)
 void CGUIImage::AllocateOnDemand()
 {
   // if we're hidden, we can free our resources and return
-  if (!IsVisible() && m_visible != DELAYED)
+  if (!IsVisible() && m_visible != DELAYED && m_bDynamicResourceAlloc)
   {
-    if (m_bDynamicResourceAlloc && m_texture->IsAllocated())
-      FreeResourcesButNotAnims();
+    FreeResourcesButNotAnims();
     return;
   }
 
   // either visible or delayed - we need the resources allocated in either case
-  if (!m_texture->IsAllocated())
+  if (!m_textureCurrent->IsAllocated())
     AllocResources();
 }
 
 void CGUIImage::Process(unsigned int currentTime, CDirtyRegionList &dirtyregions)
 {
-  // check whether our image failed to allocate, and if so drop back to the fallback image
-  if (m_texture->FailedToAlloc() && m_texture->GetFileName() != m_info.GetFallback())
-  {
-    if (!m_currentFallback.empty() && m_texture->GetFileName() != m_currentFallback)
-      m_texture->SetFileName(m_currentFallback);
-    else
-      m_texture->SetFileName(m_info.GetFallback());
-  }
+  ProcessState();
 
-  if (m_crossFadeTime)
-  {
-    // make sure our texture has started allocating
-    if (m_texture->AllocResources())
-      MarkDirtyRegion();
+  ProcessAllocation();
 
-    // compute the frame time
-    unsigned int frameTime = 0;
-    if (m_lastRenderTime)
-      frameTime = currentTime - m_lastRenderTime;
-    if (!frameTime)
-      frameTime = (unsigned int)(1000 / CServiceBroker::GetWinSystem()->GetGfxContext().GetFPS());
-    m_lastRenderTime = currentTime;
+  if (!m_isTransitioning ||
+      (!m_textureNext->ReadyToRender() && !m_textureNext->GetFileName().empty()))
+    ProcessNoTransition(currentTime);
+  else if (!m_crossFadeTime)
+    ProcessInstantTransition(currentTime);
+  else
+    ProcessFadingTransition(currentTime);
 
-    if (m_fadingTextures.size())  // have some fading images
-    { // anything other than the last old texture needs to be faded out as per usual
-      for (std::vector<CFadingTexture *>::iterator i = m_fadingTextures.begin(); i != m_fadingTextures.end() - 1;)
-      {
-        if (!ProcessFading(*i, frameTime, currentTime))
-          i = m_fadingTextures.erase(i);
-        else
-          ++i;
-      }
-
-      if (m_texture->ReadyToRender() || m_texture->GetFileName().empty())
-      { // fade out the last one as well
-        if (!ProcessFading(m_fadingTextures[m_fadingTextures.size() - 1], frameTime, currentTime))
-          m_fadingTextures.erase(m_fadingTextures.end() - 1);
-      }
-      else
-      { // keep the last one fading in
-        CFadingTexture *texture = m_fadingTextures[m_fadingTextures.size() - 1];
-        texture->m_fadeTime += frameTime;
-        if (texture->m_fadeTime > m_crossFadeTime)
-          texture->m_fadeTime = m_crossFadeTime;
-
-        if (texture->m_texture->SetAlpha(GetFadeLevel(texture->m_fadeTime)))
-          MarkDirtyRegion();
-        if (texture->m_texture->SetDiffuseColor(m_diffuseColor))
-          MarkDirtyRegion();
-        if (texture->m_texture->Process(currentTime))
-          MarkDirtyRegion();
-      }
-    }
-
-    if (m_texture->ReadyToRender() || m_texture->GetFileName().empty())
-    { // fade the new one in
-      m_currentFadeTime += frameTime;
-      if (m_currentFadeTime > m_crossFadeTime || frameTime == 0) // for if we allocate straight away on creation
-        m_currentFadeTime = m_crossFadeTime;
-    }
-    if (m_texture->SetAlpha(GetFadeLevel(m_currentFadeTime)))
-      MarkDirtyRegion();
-  }
-
-  if (!m_texture->GetDiffuseColor().HasInfo())
+  if (!m_textureCurrent->GetDiffuseColor().HasInfo())
     UpdateDiffuseColor(nullptr);
 
-  if (m_texture->Process(currentTime))
-    MarkDirtyRegion();
-
   CGUIControl::Process(currentTime, dirtyregions);
 }
 
-void CGUIImage::Render()
+void CGUIImage::ProcessState()
 {
-  if (!IsVisible()) return;
+  if (!m_hasNewStagingTexture)
+    return;
 
-  for (auto& itr : m_fadingTextures)
-    itr->m_texture->Render();
+  std::string fileName = m_nameStaging;
+  if (fileName.empty())
+    fileName = GetFallback(fileName);
 
-  m_texture->Render();
+  if (m_nameCurrent == fileName || m_textureCurrent->GetFileName() == fileName)
+  {
+    // the current texture might be a fallback from a image which failed to
+    // load, and it might be the texture we want.
+    if (m_textureCurrent->GetFileName() == fileName && m_nameCurrent != fileName)
+      m_nameCurrent = fileName;
 
-  CGUIControl::Render();
+    if (m_isTransitioning)
+    {
+      if (m_textureNext->ReadyToRender() || m_textureNext->GetFileName().empty())
+      {
+        // if the current texture (which is fading out) is our desired texture,
+        // reverse the animation.
+        std::swap(m_textureCurrent, m_textureNext);
+        std::swap(m_nameCurrent, m_nameNext);
+        m_currentFadeTime = m_crossFadeTime - m_currentFadeTime;
+      }
+      else
+      {
+        // if we are about to fade but the new texture is not ready, we want to
+        // keep the current texture, and cancel the new texture.
+        m_isTransitioning = false;
+        m_textureNext->SetFileName("");
+        m_nameNext = "";
+      }
+    }
+
+    m_hasNewStagingTexture = false;
+    return;
+  }
+
+  // our fading-in texture is already set
+  if (m_nameNext == fileName || m_textureNext->GetFileName() == fileName)
+  {
+    // the next texture might be a fallback from a image which failed to load,
+    // and it might be the texture we want.
+    if (m_textureNext->GetFileName() == fileName && m_nameNext != fileName)
+      m_nameNext = fileName;
+
+    // ensure that the transition is on its way.
+    if (!m_isTransitioning &&
+        (m_textureNext->ReadyToRender() || m_textureNext->GetFileName().empty()))
+      m_isTransitioning = true;
+
+    m_hasNewStagingTexture = false;
+    return;
+  }
+
+  // can't set new texture during animation.
+  if (m_isTransitioning && (m_textureNext->ReadyToRender() || m_textureNext->GetFileName().empty()))
+    return;
+
+  // finally, we can request a new image.
+  m_textureNext->SetFileName(fileName);
+  m_nameNext = fileName;
+  m_isTransitioning = true;
+  m_hasNewStagingTexture = false;
 }
 
-bool CGUIImage::ProcessFading(CGUIImage::CFadingTexture *texture, unsigned int frameTime, unsigned int currentTime)
+void CGUIImage::ProcessAllocation()
 {
-  assert(texture);
-  if (texture->m_fadeTime <= frameTime)
-  { // time to kill off the texture
-    MarkDirtyRegion();
-    delete texture;
-    return false;
+  m_textureCurrent->AllocResources();
+  m_textureNext->AllocResources();
+
+  if (m_isTransitioning && m_textureNext->FailedToAlloc())
+  {
+    if (m_textureNext->GetFileName() != m_info.GetFallback())
+      m_textureNext->SetFileName(GetFallback(m_nameNext));
+    else
+      m_textureNext->SetFileName("");
+
+    m_textureNext->AllocResources();
   }
-  // render this texture
-  texture->m_fadeTime -= frameTime;
 
-  if (texture->m_texture->SetAlpha(GetFadeLevel(texture->m_fadeTime)))
-    MarkDirtyRegion();
-  if (texture->m_texture->SetDiffuseColor(m_diffuseColor))
-    MarkDirtyRegion();
-  if (texture->m_texture->Process(currentTime))
-    MarkDirtyRegion();
+  if (m_textureCurrent->FailedToAlloc())
+  {
+    if (m_textureCurrent->GetFileName() != m_info.GetFallback())
+      m_textureCurrent->SetFileName(GetFallback(m_nameCurrent));
+    else
+      m_textureCurrent->SetFileName("");
 
-  return true;
+    m_textureCurrent->AllocResources();
+  }
+}
+
+void CGUIImage::ProcessNoTransition(unsigned int currentTime)
+{
+  if (m_textureCurrent->Process(currentTime))
+    MarkDirtyRegion();
+}
+
+void CGUIImage::ProcessInstantTransition(unsigned int currentTime)
+{
+  std::swap(m_textureCurrent, m_textureNext);
+  std::swap(m_nameCurrent, m_nameNext);
+
+  m_nameNext = "";
+  m_textureNext->SetFileName("");
+
+  m_textureCurrent->Process(currentTime);
+
+  m_isTransitioning = false;
+
+  MarkDirtyRegion();
+}
+
+void CGUIImage::ProcessFadingTransition(unsigned int currentTime)
+{
+  unsigned int frameTime = 0;
+  if (m_lastRenderTime)
+    frameTime = currentTime - m_lastRenderTime;
+  if (!frameTime)
+    frameTime = (unsigned int)(1000 / CServiceBroker::GetWinSystem()->GetGfxContext().GetFPS());
+  m_lastRenderTime = currentTime;
+
+  m_currentFadeTime += frameTime;
+  if (m_currentFadeTime > m_crossFadeTime ||
+      frameTime == 0) // for if we allocate straight away on creation
+    m_currentFadeTime = m_crossFadeTime;
+
+  if (m_currentFadeTime < m_crossFadeTime)
+  {
+    m_textureCurrent->SetAlpha(GetFadeLevel(m_crossFadeTime - m_currentFadeTime));
+
+    m_textureNext->SetAlpha(GetFadeLevel(m_currentFadeTime));
+    m_textureNext->Process(currentTime);
+  }
+  else
+  {
+    std::swap(m_textureCurrent, m_textureNext);
+    std::swap(m_nameCurrent, m_nameNext);
+
+    m_textureCurrent->SetAlpha(0xff);
+
+    m_nameNext = "";
+    m_textureNext->SetFileName("");
+
+    m_currentFadeTime = 0;
+    m_lastRenderTime = 0;
+    m_isTransitioning = false;
+  }
+
+  m_textureCurrent->Process(currentTime);
+
+  MarkDirtyRegion();
+}
+
+void CGUIImage::Render()
+{
+  if (!IsVisible())
+    return;
+
+  if (m_isTransitioning)
+    m_textureNext->Render();
+
+  m_textureCurrent->Render();
+
+  CGUIControl::Render();
 }
 
 bool CGUIImage::OnAction(const CAction &action)
@@ -239,22 +318,29 @@ bool CGUIImage::OnMessage(CGUIMessage& message)
 
 void CGUIImage::AllocResources()
 {
-  if (m_texture->GetFileName().empty())
+  if (m_textureCurrent->GetFileName().empty())
     return;
 
   CGUIControl::AllocResources();
-  m_texture->AllocResources();
+  m_textureCurrent->AllocResources();
 }
 
 void CGUIImage::FreeTextures(bool immediately /* = false */)
 {
-  m_texture->FreeResources(immediately);
-  for (unsigned int i = 0; i < m_fadingTextures.size(); i++)
-    delete m_fadingTextures[i];
-  m_fadingTextures.clear();
-  m_currentTexture.clear();
+  m_textureNext->FreeResources(immediately);
+  m_textureNext->SetFileName("");
+  m_nameNext = "";
+
+  m_textureCurrent->FreeResources(immediately);
   if (!m_info.IsConstant()) // constant textures never change
-    m_texture->SetFileName("");
+  {
+    m_textureCurrent->SetFileName("");
+    m_nameCurrent = "";
+  }
+
+  m_isTransitioning = false;
+  m_lastRenderTime = 0;
+  m_currentFadeTime = 0;
 }
 
 void CGUIImage::FreeResources(bool immediately)
@@ -265,7 +351,8 @@ void CGUIImage::FreeResources(bool immediately)
 
 void CGUIImage::SetInvalid()
 {
-  m_texture->SetInvalid();
+  m_textureCurrent->SetInvalid();
+  m_textureNext->SetInvalid();
   CGUIControl::SetInvalid();
 }
 
@@ -281,7 +368,8 @@ void CGUIImage::FreeResourcesButNotAnims()
 void CGUIImage::DynamicResourceAlloc(bool bOnOff)
 {
   m_bDynamicResourceAlloc = bOnOff;
-  m_texture->DynamicResourceAlloc(bOnOff);
+  m_textureCurrent->DynamicResourceAlloc(bOnOff);
+  m_textureNext->DynamicResourceAlloc(bOnOff);
   CGUIControl::DynamicResourceAlloc(bOnOff);
 }
 
@@ -292,99 +380,80 @@ bool CGUIImage::CanFocus() const
 
 float CGUIImage::GetTextureWidth() const
 {
-  return m_texture->GetTextureWidth();
+  return m_textureCurrent->GetTextureWidth();
 }
 
 float CGUIImage::GetTextureHeight() const
 {
-  return m_texture->GetTextureHeight();
+  return m_textureCurrent->GetTextureHeight();
 }
 
 CRect CGUIImage::CalcRenderRegion() const
 {
-  CRect region = m_texture->GetRenderRect();
+  CRect region = m_textureCurrent->GetRenderRect();
 
-  for (const auto& itr : m_fadingTextures)
-    region.Union(itr->m_texture->GetRenderRect());
+  if (m_isTransitioning && m_textureNext->ReadyToRender())
+    region.Union(m_textureNext->GetRenderRect());
 
   return CGUIControl::CalcRenderRegion().Intersect(region);
 }
 
 const std::string &CGUIImage::GetFileName() const
 {
-  return m_texture->GetFileName();
+  return m_textureCurrent->GetFileName();
 }
 
 void CGUIImage::SetAspectRatio(const CAspectRatio &aspect)
 {
-  m_texture->SetAspectRatio(aspect);
+  m_textureCurrent->SetAspectRatio(aspect);
+  m_textureNext->SetAspectRatio(aspect);
 }
 
 void CGUIImage::SetCrossFade(unsigned int time)
 {
   m_crossFadeTime = time;
-  if (!m_crossFadeTime && m_texture->IsLazyLoaded() && !m_info.GetFallback().empty())
-    m_crossFadeTime = 1;
 }
 
 void CGUIImage::SetFileName(const std::string& strFileName, bool setConstant, const bool useCache)
 {
   if (setConstant)
     m_info.SetLabel(strFileName, "", GetParentID());
-
-  // Set whether or not to use cache
-  m_texture->SetUseCache(useCache);
-
-  if (m_crossFadeTime)
-  {
-    // set filename on the next texture
-    if (m_currentTexture == strFileName)
-      return; // nothing to do - we already have this image
-
-    if (m_texture->ReadyToRender() || m_texture->GetFileName().empty())
-    { // save the current image
-      m_fadingTextures.push_back(new CFadingTexture(m_texture.get(), m_currentFadeTime));
-      MarkDirtyRegion();
-    }
-    m_currentFadeTime = 0;
-  }
-  if (m_currentTexture != strFileName)
-  { // texture is changing - attempt to load it, and save the name in m_currentTexture.
-    // we'll check whether it loaded or not in Render()
-    m_currentTexture = strFileName;
-    if (m_texture->SetFileName(m_currentTexture))
-      MarkDirtyRegion();
-  }
+  m_nameStaging = strFileName;
+  m_hasNewStagingTexture = true;
 }
 
 #ifdef _DEBUG
 void CGUIImage::DumpTextureUse()
 {
-  if (m_texture->IsAllocated())
+  if (m_textureCurrent->IsAllocated())
   {
     if (GetID())
-      CLog::Log(LOGDEBUG, "Image control {} using texture {}", GetID(), m_texture->GetFileName());
+      CLog::Log(LOGDEBUG, "Image control {} using texture {}", GetID(),
+                m_textureCurrent->GetFileName());
     else
-      CLog::Log(LOGDEBUG, "Using texture {}", m_texture->GetFileName());
+      CLog::Log(LOGDEBUG, "Using texture {}", m_textureCurrent->GetFileName());
   }
 }
 #endif
 
 void CGUIImage::SetWidth(float width)
 {
-  m_texture->SetWidth(width);
-  CGUIControl::SetWidth(m_texture->GetWidth());
+  m_textureCurrent->SetWidth(width);
+  m_textureNext->SetWidth(width);
+  CGUIControl::SetWidth(m_textureCurrent->GetWidth());
 }
 
 void CGUIImage::SetHeight(float height)
 {
-  m_texture->SetHeight(height);
-  CGUIControl::SetHeight(m_texture->GetHeight());
+  m_textureCurrent->SetHeight(height);
+  m_textureNext->SetHeight(height);
+  CGUIControl::SetHeight(m_textureCurrent->GetHeight());
 }
 
 void CGUIImage::SetPosition(float posX, float posY)
 {
-  m_texture->SetPosition(posX, posY);
+  m_textureCurrent->SetPosition(posX, posY);
+  m_textureNext->SetPosition(posX, posY);
   CGUIControl::SetPosition(posX, posY);
 }
 
@@ -393,7 +462,10 @@ void CGUIImage::SetInfo(const GUIINFO::CGUIInfoLabel &info)
   m_info = info;
   // a constant image never needs updating
   if (m_info.IsConstant())
-    m_texture->SetFileName(m_info.GetLabel(0));
+  {
+    m_textureCurrent->SetFileName(m_info.GetLabel(0));
+    m_nameCurrent = m_info.GetLabel(0);
+  }
 }
 
 unsigned char CGUIImage::GetFadeLevel(unsigned int time) const
@@ -410,6 +482,14 @@ unsigned char CGUIImage::GetFadeLevel(unsigned int time) const
   return (unsigned char)(255.0f * (1 - pow(1-alpha, amount))/alpha);
 }
 
+std::string CGUIImage::GetFallback(const std::string& currentName)
+{
+  if (!m_currentFallback.empty() && currentName != m_currentFallback)
+    return m_currentFallback;
+  else
+    return m_info.GetFallback();
+}
+
 std::string CGUIImage::GetDescription(void) const
 {
   return GetFileName();
diff --git a/xbmc/guilib/GUIImage.h b/xbmc/guilib/GUIImage.h
index e9e934767..4e93a9668 100644
--- a/xbmc/guilib/GUIImage.h
+++ b/xbmc/guilib/GUIImage.h
@@ -93,7 +93,12 @@ protected:
   virtual void FreeTextures(bool immediately = false);
   void FreeResourcesButNotAnims();
   unsigned char GetFadeLevel(unsigned int time) const;
-  bool ProcessFading(CFadingTexture *texture, unsigned int frameTime, unsigned int currentTime);
+  std::string GetFallback(const std::string& currentName);
+  void ProcessState();
+  void ProcessAllocation();
+  void ProcessNoTransition(unsigned int currentTime);
+  void ProcessInstantTransition(unsigned int currentTime);
+  void ProcessFadingTransition(unsigned int currentTime);
 
   /*!
    * \brief Update the diffuse color based on the current item infos
@@ -107,9 +112,16 @@ protected:
   CTextureInfo m_image;
   KODI::GUILIB::GUIINFO::CGUIInfoLabel m_info;
 
-  std::unique_ptr<CGUITexture> m_texture;
-  std::vector<CFadingTexture *> m_fadingTextures;
-  std::string m_currentTexture;
+  bool m_isTransitioning{false};
+  bool m_hasNewStagingTexture{false};
+
+  std::unique_ptr<CGUITexture> m_textureCurrent;
+  std::unique_ptr<CGUITexture> m_textureNext;
+
+  std::string m_nameCurrent{};
+  std::string m_nameNext{};
+  std::string m_nameStaging{};
+
   std::string m_currentFallback;
 
   unsigned int m_crossFadeTime;
diff --git a/xbmc/guilib/GUIProgressControl.cpp b/xbmc/guilib/GUIProgressControl.cpp
index 1167d64e4..09cf8916a 100644
--- a/xbmc/guilib/GUIProgressControl.cpp
+++ b/xbmc/guilib/GUIProgressControl.cpp
@@ -85,7 +85,12 @@ void CGUIProgressControl::Render()
 {
   if (!IsDisabled())
   {
-    m_guiBackground->Render();
+    const bool renderFrontToBack =
+        CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderOrder() ==
+        RENDER_ORDER_FRONT_TO_BACK;
+
+    if (!renderFrontToBack)
+      m_guiBackground->Render(-1);
 
     if (m_guiLeft->GetFileName().empty() && m_guiRight->GetFileName().empty())
     {
@@ -116,6 +121,9 @@ void CGUIProgressControl::Render()
       m_guiRight->Render();
     }
 
+    if (renderFrontToBack)
+      m_guiBackground->Render(-1);
+
     m_guiOverlay->Render();
   }
 
diff --git a/xbmc/guilib/GUISliderControl.cpp b/xbmc/guilib/GUISliderControl.cpp
index 8158f1751..a146c5be6 100644
--- a/xbmc/guilib/GUISliderControl.cpp
+++ b/xbmc/guilib/GUISliderControl.cpp
@@ -218,10 +218,19 @@ bool CGUISliderControl::ProcessSelector(CGUITexture* background,
 
 void CGUISliderControl::Render()
 {
-  if (!IsDisabled())
-    m_guiBackground->Render();
-  else
-    m_guiBackgroundDisabled->Render();
+  const bool renderFrontToBack = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderOrder() ==
+                                 RENDER_ORDER_FRONT_TO_BACK;
+
+  auto renderBackground = [this]
+  {
+    if (!IsDisabled())
+      m_guiBackground->Render(-1);
+    else
+      m_guiBackgroundDisabled->Render(-1);
+  };
+
+  if (!renderFrontToBack)
+    renderBackground();
 
   CGUITexture* nibLower = nullptr;
   if (IsActive() && m_bHasFocus && !IsDisabled() && m_currentSelector == RangeSelectorLower)
@@ -242,6 +251,10 @@ void CGUISliderControl::Render()
       nibUpper = m_guiSelectorUpperDisabled.get();
     nibUpper->Render();
   }
+
+  if (renderFrontToBack)
+    renderBackground();
+
   CGUIControl::Render();
 }
 
diff --git a/xbmc/guilib/Texture.cpp b/xbmc/guilib/Texture.cpp
index cfc506d48..91b9db90d 100644
--- a/xbmc/guilib/Texture.cpp
+++ b/xbmc/guilib/Texture.cpp
@@ -15,10 +15,12 @@
 #include "filesystem/File.h"
 #include "filesystem/ResourceFile.h"
 #include "filesystem/XbtFile.h"
+#include "guilib/TextureBase.h"
 #include "guilib/iimage.h"
 #include "guilib/imagefactory.h"
 #include "utils/URIUtils.h"
 #include "utils/log.h"
+#include "windowing/WinSystem.h"
 #if defined(TARGET_DARWIN_EMBEDDED)
 #include <ImageIO/ImageIO.h>
 #include "filesystem/File.h"
@@ -50,70 +52,6 @@ CTexture::~CTexture()
   m_pixels = NULL;
 }
 
-void CTexture::Allocate(unsigned int width, unsigned int height, XB_FMT format)
-{
-  m_imageWidth = m_originalWidth = width;
-  m_imageHeight = m_originalHeight = height;
-  m_format = format;
-  m_orientation = 0;
-
-  m_textureWidth = m_imageWidth;
-  m_textureHeight = m_imageHeight;
-
-  if (m_format & XB_FMT_DXT_MASK)
-  {
-    while (GetPitch() < CServiceBroker::GetRenderSystem()->GetMinDXTPitch())
-      m_textureWidth += GetBlockSize();
-  }
-
-  if (!CServiceBroker::GetRenderSystem()->SupportsNPOT((m_format & XB_FMT_DXT_MASK) != 0))
-  {
-    m_textureWidth = PadPow2(m_textureWidth);
-    m_textureHeight = PadPow2(m_textureHeight);
-  }
-
-  if (m_format & XB_FMT_DXT_MASK)
-  {
-    // DXT textures must be a multiple of 4 in width and height
-    m_textureWidth = ((m_textureWidth + 3) / 4) * 4;
-    m_textureHeight = ((m_textureHeight + 3) / 4) * 4;
-  }
-  else
-  {
-    // align all textures so that they have an even width
-    // in some circumstances when we downsize a thumbnail
-    // which has an uneven number of pixels in width
-    // we crash in CPicture::ScaleImage in ffmpegs swscale
-    // because it tries to access beyond the source memory
-    // (happens on osx and ios)
-    // UPDATE: don't just update to be on an even width;
-    // ffmpegs swscale relies on a 16-byte stride on some systems
-    // so the textureWidth needs to be a multiple of 16. see ffmpeg
-    // swscale headers for more info.
-    m_textureWidth = ((m_textureWidth + 15) / 16) * 16;
-  }
-
-  // check for max texture size
-  #define CLAMP(x, y) { if (x > y) x = y; }
-  CLAMP(m_textureWidth, CServiceBroker::GetRenderSystem()->GetMaxTextureSize());
-  CLAMP(m_textureHeight, CServiceBroker::GetRenderSystem()->GetMaxTextureSize());
-  CLAMP(m_imageWidth, m_textureWidth);
-  CLAMP(m_imageHeight, m_textureHeight);
-
-  KODI::MEMORY::AlignedFree(m_pixels);
-  m_pixels = NULL;
-  if (GetPitch() * GetRows() > 0)
-  {
-    size_t size = GetPitch() * GetRows();
-    m_pixels = static_cast<unsigned char*>(KODI::MEMORY::AlignedMalloc(size, 32));
-
-    if (m_pixels == nullptr)
-    {
-      CLog::Log(LOGERROR, "{} - Could not allocate {} bytes. Out of memory.", __FUNCTION__, size);
-    }
-  }
-}
-
 void CTexture::Update(unsigned int width,
                       unsigned int height,
                       unsigned int pitch,
@@ -156,39 +94,6 @@ void CTexture::Update(unsigned int width,
     LoadToGPU();
 }
 
-void CTexture::ClampToEdge()
-{
-  if (m_pixels == nullptr)
-    return;
-
-  unsigned int imagePitch = GetPitch(m_imageWidth);
-  unsigned int imageRows = GetRows(m_imageHeight);
-  unsigned int texturePitch = GetPitch(m_textureWidth);
-  unsigned int textureRows = GetRows(m_textureHeight);
-  if (imagePitch < texturePitch)
-  {
-    unsigned int blockSize = GetBlockSize();
-    unsigned char *src = m_pixels + imagePitch - blockSize;
-    unsigned char *dst = m_pixels;
-    for (unsigned int y = 0; y < imageRows; y++)
-    {
-      for (unsigned int x = imagePitch; x < texturePitch; x += blockSize)
-        memcpy(dst + x, src, blockSize);
-      dst += texturePitch;
-    }
-  }
-
-  if (imageRows < textureRows)
-  {
-    unsigned char *dst = m_pixels + imageRows * texturePitch;
-    for (unsigned int y = imageRows; y < textureRows; y++)
-    {
-      memcpy(dst, dst - texturePitch, texturePitch);
-      dst += texturePitch;
-    }
-  }
-}
-
 std::unique_ptr<CTexture> CTexture::LoadFromFile(const std::string& texturePath,
                                                  unsigned int idealWidth,
                                                  unsigned int idealHeight,
@@ -332,26 +237,51 @@ bool CTexture::LoadIImage(IImage* pImage,
                           unsigned int width,
                           unsigned int height)
 {
-  if(pImage != NULL && pImage->LoadImageFromMemory(buffer, bufSize, width, height))
-  {
-    if (pImage->Width() > 0 && pImage->Height() > 0)
-    {
-      Allocate(pImage->Width(), pImage->Height(), XB_FMT_A8R8G8B8);
-      if (m_pixels != nullptr && pImage->Decode(m_pixels, GetTextureWidth(), GetRows(), GetPitch(), XB_FMT_A8R8G8B8))
-      {
-        if (pImage->Orientation())
-          m_orientation = pImage->Orientation() - 1;
-        m_hasAlpha = pImage->hasAlpha();
-        m_originalWidth = pImage->originalWidth();
-        m_originalHeight = pImage->originalHeight();
-        m_imageWidth = pImage->Width();
-        m_imageHeight = pImage->Height();
-        ClampToEdge();
-        return true;
-      }
-    }
-  }
-  return false;
+  if (pImage == nullptr)
+    return false;
+
+  if (!pImage->LoadImageFromMemory(buffer, bufSize, width, height))
+    return false;
+
+  if (pImage->Width() == 0 || pImage->Height() == 0)
+    return false;
+
+  Allocate(pImage->Width(), pImage->Height(), XB_FMT_A8R8G8B8);
+
+  if (m_pixels == nullptr)
+    return false;
+
+  if (!pImage->Decode(m_pixels, GetTextureWidth(), GetRows(), GetPitch(), XB_FMT_A8R8G8B8))
+    return false;
+
+  if (pImage->Orientation())
+    m_orientation = pImage->Orientation() - 1;
+
+  m_hasAlpha = pImage->hasAlpha();
+  m_originalWidth = pImage->originalWidth();
+  m_originalHeight = pImage->originalHeight();
+  m_imageWidth = pImage->Width();
+  m_imageHeight = pImage->Height();
+
+  ClampToEdge();
+
+  return true;
+}
+
+void CTexture::LoadToGPUAsync()
+{
+  // Already in main context?
+  if (CServiceBroker::GetWinSystem()->HasContext())
+    return;
+
+  if (!CServiceBroker::GetWinSystem()->BindTextureUploadContext())
+    return;
+
+  LoadToGPU();
+
+  SyncGPU();
+
+  CServiceBroker::GetWinSystem()->UnbindTextureUploadContext();
 }
 
 bool CTexture::LoadFromMemory(unsigned int width,
@@ -397,99 +327,3 @@ bool CTexture::LoadPaletted(unsigned int width,
   ClampToEdge();
   return true;
 }
-
-unsigned int CTexture::PadPow2(unsigned int x)
-{
-  --x;
-  x |= x >> 1;
-  x |= x >> 2;
-  x |= x >> 4;
-  x |= x >> 8;
-  x |= x >> 16;
-  return ++x;
-}
-
-bool CTexture::SwapBlueRed(unsigned char* pixels,
-                           unsigned int height,
-                           unsigned int pitch,
-                           unsigned int elements,
-                           unsigned int offset)
-{
-  if (!pixels) return false;
-  unsigned char *dst = pixels;
-  for (unsigned int y = 0; y < height; y++)
-  {
-    dst = pixels + (y * pitch);
-    for (unsigned int x = 0; x < pitch; x+=elements)
-      std::swap(dst[x+offset], dst[x+2+offset]);
-  }
-  return true;
-}
-
-unsigned int CTexture::GetPitch(unsigned int width) const
-{
-  switch (m_format)
-  {
-  case XB_FMT_DXT1:
-    return ((width + 3) / 4) * 8;
-  case XB_FMT_DXT3:
-  case XB_FMT_DXT5:
-  case XB_FMT_DXT5_YCoCg:
-    return ((width + 3) / 4) * 16;
-  case XB_FMT_A8:
-    return width;
-  case XB_FMT_RGB8:
-    return (((width + 1)* 3 / 4) * 4);
-  case XB_FMT_RGBA8:
-  case XB_FMT_A8R8G8B8:
-  default:
-    return width*4;
-  }
-}
-
-unsigned int CTexture::GetRows(unsigned int height) const
-{
-  switch (m_format)
-  {
-  case XB_FMT_DXT1:
-    return (height + 3) / 4;
-  case XB_FMT_DXT3:
-  case XB_FMT_DXT5:
-  case XB_FMT_DXT5_YCoCg:
-    return (height + 3) / 4;
-  default:
-    return height;
-  }
-}
-
-unsigned int CTexture::GetBlockSize() const
-{
-  switch (m_format)
-  {
-  case XB_FMT_DXT1:
-    return 8;
-  case XB_FMT_DXT3:
-  case XB_FMT_DXT5:
-  case XB_FMT_DXT5_YCoCg:
-    return 16;
-  case XB_FMT_A8:
-    return 1;
-  default:
-    return 4;
-  }
-}
-
-bool CTexture::HasAlpha() const
-{
-  return m_hasAlpha;
-}
-
-void CTexture::SetMipmapping()
-{
-  m_mipmapping = true;
-}
-
-bool CTexture::IsMipmapped() const
-{
-  return m_mipmapping;
-}
diff --git a/xbmc/guilib/Texture.h b/xbmc/guilib/Texture.h
index 5e32e60f8..6e5a4ff1a 100644
--- a/xbmc/guilib/Texture.h
+++ b/xbmc/guilib/Texture.h
@@ -8,6 +8,7 @@
 
 #pragma once
 
+#include "guilib/TextureBase.h"
 #include "guilib/TextureFormats.h"
 
 #include <cstddef>
@@ -21,17 +22,11 @@ class IImage;
 struct COLOR {unsigned char b,g,r,x;};	// Windows GDI expects 4bytes per color
 #pragma pack()
 
-enum class TEXTURE_SCALING
-{
-  LINEAR,
-  NEAREST,
-};
-
 /*!
 \ingroup textures
-\brief Base texture class, subclasses of which depend on the render spec (DX, GL etc.)
+\brief Texture loader class, subclasses of which depend on the render spec (DX, GL etc.)
 */
-class CTexture
+class CTexture : public CTextureBase
 {
 
 public:
@@ -86,47 +81,26 @@ public:
                     const unsigned char* pixels,
                     const COLOR* palette);
 
-  bool HasAlpha() const;
-  void SetAlpha(bool hasAlpha) { m_hasAlpha = hasAlpha; }
-
-  void SetMipmapping();
-  bool IsMipmapped() const;
-  void SetScalingMethod(TEXTURE_SCALING scalingMethod) { m_scalingMethod = scalingMethod; }
-  TEXTURE_SCALING GetScalingMethod() const { return m_scalingMethod; }
-  void SetCacheMemory(bool bCacheMemory) { m_bCacheMemory = bCacheMemory; }
-  bool GetCacheMemory() const { return m_bCacheMemory; }
-
-  virtual void CreateTextureObject() = 0;
-  virtual void DestroyTextureObject() = 0;
-  virtual void LoadToGPU() = 0;
-  virtual void BindToUnit(unsigned int unit) = 0;
-
-  unsigned char* GetPixels() const { return m_pixels; }
-  unsigned int GetPitch() const { return GetPitch(m_textureWidth); }
-  unsigned int GetRows() const { return GetRows(m_textureHeight); }
-  unsigned int GetTextureWidth() const { return m_textureWidth; }
-  unsigned int GetTextureHeight() const { return m_textureHeight; }
-  unsigned int GetWidth() const { return m_imageWidth; }
-  unsigned int GetHeight() const { return m_imageHeight; }
-  /*! \brief return the original width of the image, before scaling/cropping */
-  unsigned int GetOriginalWidth() const { return m_originalWidth; }
-  /*! \brief return the original height of the image, before scaling/cropping */
-  unsigned int GetOriginalHeight() const { return m_originalHeight; }
-
-  int GetOrientation() const { return m_orientation; }
-  void SetOrientation(int orientation) { m_orientation = orientation; }
-
   void Update(unsigned int width,
               unsigned int height,
               unsigned int pitch,
               XB_FMT format,
               const unsigned char* pixels,
               bool loadToGPU);
-  void Allocate(unsigned int width, unsigned int height, XB_FMT format);
-  void ClampToEdge();
 
-  static unsigned int PadPow2(unsigned int x);
-  static bool SwapBlueRed(unsigned char *pixels, unsigned int height, unsigned int pitch, unsigned int elements = 4, unsigned int offset=0);
+  /*! 
+   * \brief Uploads the texture to the GPU. 
+   */
+  void LoadToGPUAsync();
+
+  virtual void CreateTextureObject() = 0;
+  virtual void DestroyTextureObject() = 0;
+  virtual void LoadToGPU() = 0;
+  /*! 
+   * \brief Blocks execution until the previous GFX commands have been processed.
+   */
+  virtual void SyncGPU(){};
+  virtual void BindToUnit(unsigned int unit) = 0;
 
 private:
   // no copy constructor
@@ -136,25 +110,9 @@ protected:
   bool LoadFromFileInMem(unsigned char* buffer, size_t size, const std::string& mimeType,
                          unsigned int maxWidth, unsigned int maxHeight);
   bool LoadFromFileInternal(const std::string& texturePath, unsigned int maxWidth, unsigned int maxHeight, bool requirePixels, const std::string& strMimeType = "");
-  bool LoadIImage(IImage* pImage, unsigned char* buffer, unsigned int bufSize, unsigned int width, unsigned int height);
-  // helpers for computation of texture parameters for compressed textures
-  unsigned int GetPitch(unsigned int width) const;
-  unsigned int GetRows(unsigned int height) const;
-  unsigned int GetBlockSize() const;
-
-  unsigned int m_imageWidth;
-  unsigned int m_imageHeight;
-  unsigned int m_textureWidth;
-  unsigned int m_textureHeight;
-  unsigned int m_originalWidth;   ///< original image width before scaling or cropping
-  unsigned int m_originalHeight;  ///< original image height before scaling or cropping
-
-  unsigned char* m_pixels;
-  bool m_loadedToGPU;
-  XB_FMT m_format;
-  int m_orientation;
-  bool m_hasAlpha =  true ;
-  bool m_mipmapping =  false ;
-  TEXTURE_SCALING m_scalingMethod = TEXTURE_SCALING::LINEAR;
-  bool m_bCacheMemory = false;
+  bool LoadIImage(IImage* pImage,
+                  unsigned char* buffer,
+                  unsigned int bufSize,
+                  unsigned int width,
+                  unsigned int height);
 };
diff --git a/xbmc/guilib/TextureBase.cpp b/xbmc/guilib/TextureBase.cpp
new file mode 100644
index 000000000..63c510ff9
--- /dev/null
+++ b/xbmc/guilib/TextureBase.cpp
@@ -0,0 +1,414 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "TextureBase.h"
+
+#include "ServiceBroker.h"
+#include "commons/ilog.h"
+#include "guilib/TextureFormats.h"
+#include "rendering/RenderSystem.h"
+#include "utils/MemUtils.h"
+#include "utils/log.h"
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+#include <exception>
+#include <utility>
+
+void CTextureBase::Allocate(uint32_t width, uint32_t height, XB_FMT format)
+{
+  SetKDFormat(format);
+  m_imageWidth = m_originalWidth = width;
+  m_imageHeight = m_originalHeight = height;
+  m_format = format;
+  m_orientation = 0;
+
+  m_textureWidth = m_imageWidth;
+  m_textureHeight = m_imageHeight;
+
+  if (!CServiceBroker::GetRenderSystem()->SupportsNPOT((m_textureFormat & KD_TEX_FMT_TYPE_MASK) !=
+                                                       KD_TEX_FMT_S3TC))
+  {
+    m_textureWidth = PadPow2(m_textureWidth);
+    m_textureHeight = PadPow2(m_textureHeight);
+  }
+
+  if ((m_textureFormat & KD_TEX_FMT_TYPE_MASK) == KD_TEX_FMT_S3TC)
+  {
+    // DXT textures must be a multiple of 4 in width and height
+    m_textureWidth = ((m_textureWidth + 3) / 4) * 4;
+    m_textureHeight = ((m_textureHeight + 3) / 4) * 4;
+  }
+  else
+  {
+    // align all textures so that they have an even width
+    // in some circumstances when we downsize a thumbnail
+    // which has an uneven number of pixels in width
+    // we crash in CPicture::ScaleImage in ffmpegs swscale
+    // because it tries to access beyond the source memory
+    // (happens on osx and ios)
+    // UPDATE: don't just update to be on an even width;
+    // ffmpegs swscale relies on a 16-byte stride on some systems
+    // so the textureWidth needs to be a multiple of 16. see ffmpeg
+    // swscale headers for more info.
+    m_textureWidth = ((m_textureWidth + 15) / 16) * 16;
+  }
+
+  // check for max texture size
+  m_textureWidth = std::min(m_textureWidth, CServiceBroker::GetRenderSystem()->GetMaxTextureSize());
+  m_textureHeight =
+      std::min(m_textureHeight, CServiceBroker::GetRenderSystem()->GetMaxTextureSize());
+  m_imageWidth = std::min(m_imageWidth, m_textureWidth);
+  m_imageHeight = std::min(m_imageHeight, m_textureHeight);
+
+  KODI::MEMORY::AlignedFree(m_pixels);
+  m_pixels = NULL;
+
+  size_t size = GetPitch() * GetRows();
+
+  if (size == 0)
+    return;
+
+  m_pixels = static_cast<unsigned char*>(KODI::MEMORY::AlignedMalloc(size, 32));
+
+  if (m_pixels == nullptr)
+    CLog::Log(LOGERROR, "{} - Could not allocate {} bytes. Out of memory.", __FUNCTION__, size);
+}
+
+uint32_t CTextureBase::PadPow2(uint32_t x)
+{
+  --x;
+  x |= x >> 1;
+  x |= x >> 2;
+  x |= x >> 4;
+  x |= x >> 8;
+  x |= x >> 16;
+  return ++x;
+}
+
+bool CTextureBase::SwapBlueRed(
+    uint8_t* pixels, uint32_t height, uint32_t pitch, uint32_t elements, uint32_t offset)
+{
+  if (!pixels)
+    return false;
+  uint8_t* dst = pixels;
+  for (uint32_t y = 0; y < height; y++)
+  {
+    dst = pixels + (y * pitch);
+    for (uint32_t x = 0; x < pitch; x += elements)
+      std::swap(dst[x + offset], dst[x + 2 + offset]);
+  }
+  return true;
+}
+
+void CTextureBase::ClampToEdge()
+{
+  if (m_pixels == nullptr)
+    return;
+
+  uint32_t imagePitch = GetPitch(m_imageWidth);
+  uint32_t imageRows = GetRows(m_imageHeight);
+  uint32_t texturePitch = GetPitch(m_textureWidth);
+  uint32_t textureRows = GetRows(m_textureHeight);
+
+  if (imagePitch < texturePitch)
+  {
+    uint32_t blockSize = GetBlockSize();
+    unsigned char* src = m_pixels + imagePitch - blockSize;
+    unsigned char* dst = m_pixels;
+    for (uint32_t y = 0; y < imageRows; y++)
+    {
+      for (uint32_t x = imagePitch; x < texturePitch; x += blockSize)
+        memcpy(dst + x, src, blockSize);
+      dst += texturePitch;
+      src += texturePitch;
+    }
+  }
+
+  if (imageRows < textureRows)
+  {
+    unsigned char* dst = m_pixels + imageRows * texturePitch;
+    for (uint32_t y = imageRows; y < textureRows; y++)
+    {
+      memcpy(dst, dst - texturePitch, texturePitch);
+      dst += texturePitch;
+    }
+  }
+}
+
+uint32_t CTextureBase::GetPitch(uint32_t width) const
+{
+  uint32_t blockWidth = GetBlockWidth();
+  uint32_t pitch = ((width + blockWidth - 1) / blockWidth) * GetBlockSize();
+
+  // For the GPU, RGB8 needs to be aligned to 32 bit
+  if (m_textureFormat == KD_TEX_FMT_SDR_RGB8)
+    pitch = ((pitch + 3) / 4) * 4;
+
+  return pitch;
+}
+
+uint32_t CTextureBase::GetRows(uint32_t height) const
+{
+  uint32_t blockHeight = GetBlockHeight();
+  return (height + blockHeight - 1) / blockHeight;
+}
+
+uint32_t CTextureBase::GetBlockWidth() const
+{
+  switch (m_textureFormat)
+  {
+    case KD_TEX_FMT_SDR_R8:
+    case KD_TEX_FMT_SDR_RG8:
+    case KD_TEX_FMT_SDR_R5G6B5:
+    case KD_TEX_FMT_SDR_RGB5_A1:
+    case KD_TEX_FMT_SDR_RGBA4:
+    case KD_TEX_FMT_SDR_RGB8:
+    case KD_TEX_FMT_SDR_RGBA8:
+    case KD_TEX_FMT_SDR_BGRA8:
+    case KD_TEX_FMT_HDR_R16f:
+    case KD_TEX_FMT_HDR_RG16f:
+    case KD_TEX_FMT_HDR_R11F_G11F_B10F:
+    case KD_TEX_FMT_HDR_RGB9_E5:
+    case KD_TEX_FMT_HDR_RGB10_A2:
+    case KD_TEX_FMT_HDR_RGBA16f:
+      return 1;
+    case KD_TEX_FMT_YUV_YUYV8:
+      return 2;
+    case KD_TEX_FMT_S3TC_RGB8:
+    case KD_TEX_FMT_S3TC_RGB8_A1:
+    case KD_TEX_FMT_S3TC_RGB8_A4:
+    case KD_TEX_FMT_S3TC_RGBA8:
+    case KD_TEX_FMT_RGTC_R11:
+    case KD_TEX_FMT_RGTC_RG11:
+    case KD_TEX_FMT_BPTC_RGB16F:
+    case KD_TEX_FMT_BPTC_RGBA8:
+    case KD_TEX_FMT_ETC1:
+    case KD_TEX_FMT_ETC2_RGB8:
+    case KD_TEX_FMT_ETC2_RGB8_A1:
+    case KD_TEX_FMT_ETC2_RGBA8:
+    case KD_TEX_FMT_ETC2_R11:
+    case KD_TEX_FMT_ETC2_RG11:
+    case KD_TEX_FMT_ASTC_LDR_4x4:
+    case KD_TEX_FMT_ASTC_HDR_4x4:
+      return 4;
+    case KD_TEX_FMT_ASTC_LDR_5x4:
+    case KD_TEX_FMT_ASTC_LDR_5x5:
+    case KD_TEX_FMT_ASTC_HDR_5x4:
+    case KD_TEX_FMT_ASTC_HDR_5x5:
+      return 5;
+    case KD_TEX_FMT_ASTC_LDR_6x5:
+    case KD_TEX_FMT_ASTC_LDR_6x6:
+    case KD_TEX_FMT_ASTC_HDR_6x5:
+    case KD_TEX_FMT_ASTC_HDR_6x6:
+      return 6;
+    case KD_TEX_FMT_ASTC_LDR_8x5:
+    case KD_TEX_FMT_ASTC_LDR_8x6:
+    case KD_TEX_FMT_ASTC_LDR_8x8:
+    case KD_TEX_FMT_ASTC_HDR_8x5:
+    case KD_TEX_FMT_ASTC_HDR_8x6:
+    case KD_TEX_FMT_ASTC_HDR_8x8:
+      return 8;
+    case KD_TEX_FMT_ASTC_LDR_10x5:
+    case KD_TEX_FMT_ASTC_LDR_10x6:
+    case KD_TEX_FMT_ASTC_LDR_10x8:
+    case KD_TEX_FMT_ASTC_LDR_10x10:
+    case KD_TEX_FMT_ASTC_HDR_10x5:
+    case KD_TEX_FMT_ASTC_HDR_10x6:
+    case KD_TEX_FMT_ASTC_HDR_10x8:
+    case KD_TEX_FMT_ASTC_HDR_10x10:
+      return 10;
+    case KD_TEX_FMT_ASTC_LDR_12x10:
+    case KD_TEX_FMT_ASTC_LDR_12x12:
+    case KD_TEX_FMT_ASTC_HDR_12x10:
+    case KD_TEX_FMT_ASTC_HDR_12x12:
+      return 12;
+    default:
+      return 1;
+  }
+}
+
+uint32_t CTextureBase::GetBlockHeight() const
+{
+  switch (m_textureFormat)
+  {
+    case KD_TEX_FMT_SDR_R8:
+    case KD_TEX_FMT_SDR_RG8:
+    case KD_TEX_FMT_SDR_R5G6B5:
+    case KD_TEX_FMT_SDR_RGB5_A1:
+    case KD_TEX_FMT_SDR_RGBA4:
+    case KD_TEX_FMT_SDR_RGB8:
+    case KD_TEX_FMT_SDR_RGBA8:
+    case KD_TEX_FMT_SDR_BGRA8:
+    case KD_TEX_FMT_HDR_R16f:
+    case KD_TEX_FMT_HDR_RG16f:
+    case KD_TEX_FMT_HDR_R11F_G11F_B10F:
+    case KD_TEX_FMT_HDR_RGB9_E5:
+    case KD_TEX_FMT_HDR_RGB10_A2:
+    case KD_TEX_FMT_HDR_RGBA16f:
+    case KD_TEX_FMT_YUV_YUYV8:
+      return 1;
+    case KD_TEX_FMT_S3TC_RGB8:
+    case KD_TEX_FMT_S3TC_RGB8_A1:
+    case KD_TEX_FMT_S3TC_RGB8_A4:
+    case KD_TEX_FMT_S3TC_RGBA8:
+    case KD_TEX_FMT_RGTC_R11:
+    case KD_TEX_FMT_RGTC_RG11:
+    case KD_TEX_FMT_BPTC_RGB16F:
+    case KD_TEX_FMT_BPTC_RGBA8:
+    case KD_TEX_FMT_ETC1:
+    case KD_TEX_FMT_ETC2_RGB8:
+    case KD_TEX_FMT_ETC2_RGB8_A1:
+    case KD_TEX_FMT_ETC2_RGBA8:
+    case KD_TEX_FMT_ETC2_R11:
+    case KD_TEX_FMT_ETC2_RG11:
+    case KD_TEX_FMT_ASTC_LDR_4x4:
+    case KD_TEX_FMT_ASTC_LDR_5x4:
+    case KD_TEX_FMT_ASTC_HDR_4x4:
+    case KD_TEX_FMT_ASTC_HDR_5x4:
+      return 4;
+    case KD_TEX_FMT_ASTC_LDR_5x5:
+    case KD_TEX_FMT_ASTC_LDR_6x5:
+    case KD_TEX_FMT_ASTC_LDR_8x5:
+    case KD_TEX_FMT_ASTC_LDR_10x5:
+    case KD_TEX_FMT_ASTC_HDR_5x5:
+    case KD_TEX_FMT_ASTC_HDR_6x5:
+    case KD_TEX_FMT_ASTC_HDR_8x5:
+    case KD_TEX_FMT_ASTC_HDR_10x5:
+      return 5;
+    case KD_TEX_FMT_ASTC_LDR_6x6:
+    case KD_TEX_FMT_ASTC_LDR_8x6:
+    case KD_TEX_FMT_ASTC_LDR_10x6:
+    case KD_TEX_FMT_ASTC_HDR_6x6:
+    case KD_TEX_FMT_ASTC_HDR_8x6:
+    case KD_TEX_FMT_ASTC_HDR_10x6:
+      return 6;
+    case KD_TEX_FMT_ASTC_LDR_8x8:
+    case KD_TEX_FMT_ASTC_LDR_10x8:
+    case KD_TEX_FMT_ASTC_HDR_8x8:
+    case KD_TEX_FMT_ASTC_HDR_10x8:
+      return 8;
+    case KD_TEX_FMT_ASTC_LDR_10x10:
+    case KD_TEX_FMT_ASTC_LDR_12x10:
+    case KD_TEX_FMT_ASTC_HDR_10x10:
+    case KD_TEX_FMT_ASTC_HDR_12x10:
+      return 10;
+    case KD_TEX_FMT_ASTC_LDR_12x12:
+    case KD_TEX_FMT_ASTC_HDR_12x12:
+      return 12;
+    default:
+      return 4;
+  }
+}
+
+uint32_t CTextureBase::GetBlockSize() const
+{
+  switch (m_textureFormat)
+  {
+    case KD_TEX_FMT_SDR_R8:
+      return 1;
+    case KD_TEX_FMT_SDR_RG8:
+    case KD_TEX_FMT_SDR_R5G6B5:
+    case KD_TEX_FMT_SDR_RGB5_A1:
+    case KD_TEX_FMT_SDR_RGBA4:
+    case KD_TEX_FMT_HDR_R16f:
+      return 2;
+    case KD_TEX_FMT_SDR_RGB8:
+      return 3;
+    case KD_TEX_FMT_SDR_RGBA8:
+    case KD_TEX_FMT_SDR_BGRA8:
+    case KD_TEX_FMT_HDR_RG16f:
+    case KD_TEX_FMT_HDR_R11F_G11F_B10F:
+    case KD_TEX_FMT_HDR_RGB9_E5:
+    case KD_TEX_FMT_HDR_RGB10_A2:
+    case KD_TEX_FMT_YUV_YUYV8:
+      return 4;
+    case KD_TEX_FMT_HDR_RGBA16f:
+    case KD_TEX_FMT_S3TC_RGB8:
+    case KD_TEX_FMT_S3TC_RGB8_A1:
+    case KD_TEX_FMT_RGTC_R11:
+    case KD_TEX_FMT_ETC1:
+    case KD_TEX_FMT_ETC2_RGB8:
+    case KD_TEX_FMT_ETC2_RGB8_A1:
+    case KD_TEX_FMT_ETC2_R11:
+      return 8;
+    case KD_TEX_FMT_S3TC_RGB8_A4:
+    case KD_TEX_FMT_S3TC_RGBA8:
+    case KD_TEX_FMT_RGTC_RG11:
+    case KD_TEX_FMT_BPTC_RGB16F:
+    case KD_TEX_FMT_BPTC_RGBA8:
+    case KD_TEX_FMT_ETC2_RG11:
+    case KD_TEX_FMT_ETC2_RGBA8:
+    case KD_TEX_FMT_ASTC_LDR_4x4:
+    case KD_TEX_FMT_ASTC_LDR_5x4:
+    case KD_TEX_FMT_ASTC_LDR_5x5:
+    case KD_TEX_FMT_ASTC_LDR_6x5:
+    case KD_TEX_FMT_ASTC_LDR_6x6:
+    case KD_TEX_FMT_ASTC_LDR_8x5:
+    case KD_TEX_FMT_ASTC_LDR_8x6:
+    case KD_TEX_FMT_ASTC_LDR_8x8:
+    case KD_TEX_FMT_ASTC_LDR_10x5:
+    case KD_TEX_FMT_ASTC_LDR_10x6:
+    case KD_TEX_FMT_ASTC_LDR_10x8:
+    case KD_TEX_FMT_ASTC_LDR_10x10:
+    case KD_TEX_FMT_ASTC_LDR_12x10:
+    case KD_TEX_FMT_ASTC_LDR_12x12:
+    case KD_TEX_FMT_ASTC_HDR_4x4:
+    case KD_TEX_FMT_ASTC_HDR_5x4:
+    case KD_TEX_FMT_ASTC_HDR_5x5:
+    case KD_TEX_FMT_ASTC_HDR_6x5:
+    case KD_TEX_FMT_ASTC_HDR_6x6:
+    case KD_TEX_FMT_ASTC_HDR_8x5:
+    case KD_TEX_FMT_ASTC_HDR_8x6:
+    case KD_TEX_FMT_ASTC_HDR_8x8:
+    case KD_TEX_FMT_ASTC_HDR_10x5:
+    case KD_TEX_FMT_ASTC_HDR_10x6:
+    case KD_TEX_FMT_ASTC_HDR_10x8:
+    case KD_TEX_FMT_ASTC_HDR_10x10:
+    case KD_TEX_FMT_ASTC_HDR_12x10:
+    case KD_TEX_FMT_ASTC_HDR_12x12:
+      return 16;
+    default:
+      return 4;
+  }
+}
+
+void CTextureBase::SetKDFormat(XB_FMT xbFMT)
+{
+  switch (xbFMT)
+  {
+    case XB_FMT_DXT1:
+      m_textureFormat = KD_TEX_FMT_S3TC_RGB8;
+      return;
+    case XB_FMT_DXT3:
+      m_textureFormat = KD_TEX_FMT_S3TC_RGB8_A4;
+      return;
+    case XB_FMT_DXT5:
+      m_textureFormat = KD_TEX_FMT_S3TC_RGBA8;
+      return;
+    case XB_FMT_A8R8G8B8:
+      m_textureFormat = KD_TEX_FMT_SDR_BGRA8;
+      return;
+    case XB_FMT_A8:
+      m_textureFormat = KD_TEX_FMT_SDR_R8;
+      m_textureSwizzle = KD_TEX_SWIZ_111R;
+      return;
+    case XB_FMT_RGBA8:
+      m_textureFormat = KD_TEX_FMT_SDR_RGBA8;
+      return;
+    case XB_FMT_RGB8:
+      m_textureFormat = KD_TEX_FMT_SDR_RGB8;
+      return;
+    case XB_FMT_UNKNOWN:
+    case XB_FMT_DXT5_YCoCg:
+    default:
+      m_textureFormat = KD_TEX_FMT_UNKNOWN;
+      return;
+  }
+}
diff --git a/xbmc/guilib/TextureBase.h b/xbmc/guilib/TextureBase.h
new file mode 100644
index 000000000..bbabe2ae4
--- /dev/null
+++ b/xbmc/guilib/TextureBase.h
@@ -0,0 +1,117 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "guilib/TextureFormats.h"
+
+#include <memory>
+#include <string>
+
+enum class TEXTURE_SCALING
+{
+  LINEAR,
+  NEAREST,
+};
+
+/*!
+\ingroup textures
+\brief Base texture class, which holds the state of the texture, as well as all conversion functions.
+*/
+class CTextureBase
+{
+
+public:
+  CTextureBase() = default;
+  ~CTextureBase() = default;
+
+  bool HasAlpha() const { return m_hasAlpha; }
+  void SetAlpha(bool hasAlpha) { m_hasAlpha = hasAlpha; }
+
+  /*! \brief sets mipmapping. do not use in new code. will be replaced with proper scaling. */
+  void SetMipmapping() { m_mipmapping = true; }
+  /*! \brief return true if we want to mipmap */
+  bool IsMipmapped() const { return m_mipmapping; }
+
+  /*! \brief sets the scaling method. scanout systems might support more than NN/linear. */
+  void SetScalingMethod(TEXTURE_SCALING scalingMethod) { m_scalingMethod = scalingMethod; }
+  /*! \brief returns the scaling method. */
+  TEXTURE_SCALING GetScalingMethod() const { return m_scalingMethod; }
+
+  int32_t GetOrientation() const { return m_orientation; }
+  void SetOrientation(int orientation) { m_orientation = orientation; }
+
+  /*! \brief retains a shadow copy of the texture on the CPU side. */
+  void SetCacheMemory(bool bCacheMemory) { m_bCacheMemory = bCacheMemory; }
+  /*! \brief returns true if a shadow copy is kept on the CPU side. */
+  bool GetCacheMemory() const { return m_bCacheMemory; }
+
+  /*! \brief returns a pointer to the staging texture. */
+  uint8_t* GetPixels() const { return m_pixels; }
+
+  /*! \brief return the size of one row in bytes. */
+  uint32_t GetPitch() const { return GetPitch(m_textureWidth); }
+  /*! \brief return the number of rows (number of blocks in the Y direction). */
+  uint32_t GetRows() const { return GetRows(m_textureHeight); }
+  /*! \brief return the total width of the texture, which might be scaled to fit its displayed size. */
+  uint32_t GetTextureWidth() const { return m_textureWidth; }
+  /*! \brief return the total height of the texture, which might be scaled to fit its displayed size. */
+  uint32_t GetTextureHeight() const { return m_textureHeight; }
+  /*! \brief return the total width of "active" area, which might be equal or lower than the texture width due to alignment. */
+  uint32_t GetWidth() const { return m_imageWidth; }
+  /*! \brief return the total height of "active" area, which might be equal or lower than the texture height due to alignment. */
+  uint32_t GetHeight() const { return m_imageHeight; }
+  /*! \brief return the original width of the image, before scaling/cropping */
+  uint32_t GetOriginalWidth() const { return m_originalWidth; }
+  /*! \brief return the original height of the image, before scaling/cropping */
+  uint32_t GetOriginalHeight() const { return m_originalHeight; }
+
+  // allocates staging texture space.
+  void Allocate(uint32_t width, uint32_t height, XB_FMT format);
+  void ClampToEdge();
+
+  /*! \brief rounds to power of two. deprecated. */
+  static uint32_t PadPow2(uint32_t x);
+  /*! \brief swaps the blue/red channel. deprecated in this form. */
+  static bool SwapBlueRed(
+      uint8_t* pixels, uint32_t height, uint32_t pitch, uint32_t elements = 4, uint32_t offset = 0);
+
+protected:
+  // helpers for computation of texture parameters for compressed textures
+  uint32_t GetPitch(uint32_t width) const;
+  uint32_t GetRows(uint32_t height) const;
+  uint32_t GetBlockWidth() const;
+  uint32_t GetBlockHeight() const;
+  /*! \brief return the size of a compression block (tile) in bytes. */
+  uint32_t GetBlockSize() const;
+
+  void SetKDFormat(XB_FMT xbFMT);
+
+  uint32_t m_imageWidth{0};
+  uint32_t m_imageHeight{0};
+  uint32_t m_textureWidth{0};
+  uint32_t m_textureHeight{0};
+  uint32_t m_originalWidth{0}; ///< original image width before scaling or cropping
+  uint32_t m_originalHeight{0}; ///< original image height before scaling or cropping
+
+  uint8_t* m_pixels{nullptr};
+  bool m_loadedToGPU{false};
+
+  KD_TEX_FMT m_textureFormat{KD_TEX_FMT_UNKNOWN}; // new Kodi texture format
+  KD_TEX_SWIZ m_textureSwizzle{KD_TEX_SWIZ_RGBA};
+  KD_TEX_COL m_textureColorspace{KD_TEX_COL_REC709};
+  KD_TEX_TRANSFER m_textureTransfer{KD_TEX_TRANSFER_SRGB};
+  KD_TEX_ALPHA m_textureAlpha{KD_TEX_ALPHA_STRAIGHT};
+
+  XB_FMT m_format{XB_FMT_UNKNOWN}; // legacy XB format, deprecated
+  int32_t m_orientation{0};
+  bool m_hasAlpha{true};
+  bool m_mipmapping{false};
+  TEXTURE_SCALING m_scalingMethod{TEXTURE_SCALING::LINEAR};
+  bool m_bCacheMemory{false};
+};
diff --git a/xbmc/guilib/TextureFormats.h b/xbmc/guilib/TextureFormats.h
index 2aab71a94..5b7d87c5f 100644
--- a/xbmc/guilib/TextureFormats.h
+++ b/xbmc/guilib/TextureFormats.h
@@ -25,4 +25,159 @@ enum XB_FMT
   XB_FMT_MASK        = 0xFFFF,
   XB_FMT_OPAQUE      = 0x10000,
 };
+
+enum KD_TEX_FMT
+{
+  KD_TEX_FMT_UNKNOWN            = 0x0000,
+
+  // Legacy XB_FMT formats family
+  KD_TEX_FMT_LEGACY             = 0x0000,
+
+  // SDR texture family
+  KD_TEX_FMT_SDR                = 0x1000,
+  KD_TEX_FMT_SDR_R8             = 0x1000, // 8bpp, single channel
+  KD_TEX_FMT_SDR_RG8            = 0x1100, // 16bpp, dual channel
+  KD_TEX_FMT_SDR_R5G6B5         = 0x1200, // 16bpp, 5/6 bit per color channel
+  KD_TEX_FMT_SDR_RGB5_A1        = 0x1300, // 16bpp, 5 bit per color channel, pt-alpha
+  KD_TEX_FMT_SDR_RGBA4          = 0x1400, // 16bpp, 4 bit per channel
+  KD_TEX_FMT_SDR_RGB8           = 0x1500, // 24bpp, 8 bit per channel, no alpha (unsuitable for GPUs!)
+  KD_TEX_FMT_SDR_RGBA8          = 0x1600, // 32bpp, 8 bit per channel, RGBA order
+  KD_TEX_FMT_SDR_BGRA8          = 0x1700, // 32bpp, 8 bit per channel, BGRA order
+
+  // HDR texture family
+  KD_TEX_FMT_HDR                = 0x2000, 
+  KD_TEX_FMT_HDR_R16f           = 0x2100, // 16bpp, single channel float
+  KD_TEX_FMT_HDR_RG16f          = 0x2200, // 32bpp, dual channel float
+  KD_TEX_FMT_HDR_R11F_G11F_B10F = 0x2300, // 32bpp, 6e5/5e5 per color channel
+  KD_TEX_FMT_HDR_RGB9_E5        = 0x2400, // 32bpp, 9 bit color, shared 5 bit exponent
+  KD_TEX_FMT_HDR_RGB10_A2       = 0x2500, // 32bpp, 10 bit color, 2 bit alpha
+  KD_TEX_FMT_HDR_RGBA16f        = 0x2600, // 64bpp, four channel float
+
+  // YUV texture family
+  KD_TEX_FMT_YUV                = 0x3000,
+  KD_TEX_FMT_YUV_YUYV8          = 0x3000, // 16bpp, 4:2:2 packed 
+
+  // S3TC texture family
+  KD_TEX_FMT_S3TC               = 0x4000,
+  KD_TEX_FMT_S3TC_RGB8          = 0x4000, // 4bpp, RGB (BC1)
+  KD_TEX_FMT_S3TC_RGB8_A1       = 0x4100, // 4bpp, RGB, pt-alpha (BC1)
+  KD_TEX_FMT_S3TC_RGB8_A4       = 0x4200, // 8bpp, RGB, 4 bit alpha (BC2)
+  KD_TEX_FMT_S3TC_RGBA8         = 0x4300, // 8bpp, RGBA (BC3)
+
+  // RGTC (LATC) texture family
+  KD_TEX_FMT_RGTC               = 0x5000,
+  KD_TEX_FMT_RGTC_R11           = 0x5000, // 4bpp, single channel (BC4)
+  KD_TEX_FMT_RGTC_RG11          = 0x5100, // 8bpp, dual channel (BC5)
+
+  // BPTC texture family
+  KD_TEX_FMT_BPTC               = 0x6000,
+  KD_TEX_FMT_BPTC_RGB16F        = 0x6000, // 8bpp, HDR (BC6H float)
+  KD_TEX_FMT_BPTC_RGBA8         = 0x6100, // 8bpp, LDR (BC7 unorm)
+
+  // ETC1 texture family
+  KD_TEX_FMT_ETC1               = 0x7000,
+  KD_TEX_FMT_ETC1_RGB8          = 0x7000, // 4bpp, RGB
+
+  // ETC2 texture family
+  KD_TEX_FMT_ETC2               = 0x8000,
+  KD_TEX_FMT_ETC2_R11           = 0x8100, // 4bpp, single channel (EAC)
+  KD_TEX_FMT_ETC2_RG11          = 0x8200, // 8bpp, dual channel (EAC)
+  KD_TEX_FMT_ETC2_RGB8          = 0x8300, // 4bpp, RGB
+  KD_TEX_FMT_ETC2_RGB8_A1       = 0x8400, // 4bpp, RGB, pt-alpha
+  KD_TEX_FMT_ETC2_RGBA8         = 0x8500, // 8bpp, RGB, alpha EAC
+
+  // ASTC LDR texture family
+  // Bitrate varies from 8bpp (4x4 tile) to 0.89bpp (12x12 tile).
+  KD_TEX_FMT_ASTC_LDR           = 0x9000,
+  KD_TEX_FMT_ASTC_LDR_4x4       = 0x9000,
+  KD_TEX_FMT_ASTC_LDR_5x4       = 0x9100,
+  KD_TEX_FMT_ASTC_LDR_5x5       = 0x9200,
+  KD_TEX_FMT_ASTC_LDR_6x5       = 0x9300,
+  KD_TEX_FMT_ASTC_LDR_6x6       = 0x9400,
+  KD_TEX_FMT_ASTC_LDR_8x5       = 0x9500,
+  KD_TEX_FMT_ASTC_LDR_8x6       = 0x9600,
+  KD_TEX_FMT_ASTC_LDR_8x8       = 0x9700,
+  KD_TEX_FMT_ASTC_LDR_10x5      = 0x9800,
+  KD_TEX_FMT_ASTC_LDR_10x6      = 0x9900,
+  KD_TEX_FMT_ASTC_LDR_10x8      = 0x9A00,
+  KD_TEX_FMT_ASTC_LDR_10x10     = 0x9B00,
+  KD_TEX_FMT_ASTC_LDR_12x10     = 0x9C00,
+  KD_TEX_FMT_ASTC_LDR_12x12     = 0x9D00,
+
+  // ASTC HDR texture family
+  // Bitrate varies from 8bpp (4x4 tile) to 0.89bpp (12x12 tile).
+  KD_TEX_FMT_ASTC_HDR           = 0xA000,
+  KD_TEX_FMT_ASTC_HDR_4x4       = 0xA000,
+  KD_TEX_FMT_ASTC_HDR_5x4       = 0xA100,
+  KD_TEX_FMT_ASTC_HDR_5x5       = 0xA200,
+  KD_TEX_FMT_ASTC_HDR_6x5       = 0xA300,
+  KD_TEX_FMT_ASTC_HDR_6x6       = 0xA400,
+  KD_TEX_FMT_ASTC_HDR_8x5       = 0xA500,
+  KD_TEX_FMT_ASTC_HDR_8x6       = 0xA600,
+  KD_TEX_FMT_ASTC_HDR_8x8       = 0xA700,
+  KD_TEX_FMT_ASTC_HDR_10x5      = 0xA800,
+  KD_TEX_FMT_ASTC_HDR_10x6      = 0xA900,
+  KD_TEX_FMT_ASTC_HDR_10x8      = 0xAA00,
+  KD_TEX_FMT_ASTC_HDR_10x10     = 0xAB00,
+  KD_TEX_FMT_ASTC_HDR_12x10     = 0xAC00,
+  KD_TEX_FMT_ASTC_HDR_12x12     = 0xAD00,
+
+  KD_TEX_FMT_TYPE_MASK          = 0xF000,
+
+  KD_TEX_FMT_MASK               = 0xFFFF,
+};
+
+// Alpha handling
+enum KD_TEX_ALPHA
+{
+  KD_TEX_ALPHA_STRAIGHT         = 0x00000, // Straight (unmultiplied) alpha
+  KD_TEX_ALPHA_OPAQUE           = 0x10000, // No alpha
+  KD_TEX_ALPHA_PREMULTIPLIED    = 0x20000, // Premultiplied alpha
+
+  KD_TEX_ALPHA_MASK             = 0xF0000,
+};
+ 
+// Texture component swizzle or effect
+enum KD_TEX_SWIZ
+{
+  KD_TEX_SWIZ_RGBA              = 0x000000, // No swizzling
+  KD_TEX_SWIZ_RGB1              = 0x100000, // Normal swizzle, ignoring alpha
+  KD_TEX_SWIZ_RRR1              = 0x200000, // Luminance
+  KD_TEX_SWIZ_111R              = 0x300000, // Alpha
+  KD_TEX_SWIZ_RRRG              = 0x400000, // Luminance-Alpha
+  KD_TEX_SWIZ_RRRR              = 0x500000, // Intensity
+  KD_TEX_SWIZ_GGG1              = 0x600000, // Luminance (ETC1/BC1)
+  KD_TEX_SWIZ_111G              = 0x700000, // Alpha (ETC1/BC1)
+  KD_TEX_SWIZ_GGGA              = 0x800000, // Luminance-Alpha (BC2/BC3)
+  KD_TEX_SWIZ_GGGG              = 0x900000, // Intensity (ETC1/BC1)
+
+  KD_TEX_SWIZ_SDF               = 0xa00000, // Red channel contains a SDF
+  KD_TEX_SWIZ_RGB_SDF           = 0xb00000, // RGB8 texture with a SDF packed in the alpha channel
+  KD_TEX_SWIZ_MSDF              = 0xc00000, // Encoded MSDF in the color channels, Alpha is ignored
+
+  KD_TEX_SWIZ_MASK              = 0xF00000,
+};
+
+// Color space
+enum KD_TEX_COL
+{
+  KD_TEX_COL_REC709             = 0x0000000, // REC709/sRGB color space
+  KD_TEX_COL_REC2020            = 0x1000000, // REC2020 color space
+
+  KD_TEX_COL_MASK               = 0xF000000,
+};
+
+// Transfer function
+enum KD_TEX_TRANSFER
+{
+  KD_TEX_TRANSFER_SRGB          = 0x00000000,
+  KD_TEX_TRANSFER_REC709        = 0x10000000,
+  KD_TEX_TRANSFER_HLG           = 0x20000000,
+  KD_TEX_TRANSFER_LINEAR        = 0x30000000,
+  KD_TEX_TRANSFER_SQUARED       = 0x40000000,
+  KD_TEX_TRANSFER_PQ            = 0x50000000,
+
+  KD_TEX_TRANSFER_MASK          = 0xF0000000,
+};
+
 // clang-format on
diff --git a/xbmc/guilib/TextureGL.cpp b/xbmc/guilib/TextureGL.cpp
index a2cf428b2..e6bbdef2c 100644
--- a/xbmc/guilib/TextureGL.cpp
+++ b/xbmc/guilib/TextureGL.cpp
@@ -218,6 +218,11 @@ void CGLTexture::LoadToGPU()
   m_loadedToGPU = true;
 }
 
+void CGLTexture::SyncGPU()
+{
+  glFinish();
+}
+
 void CGLTexture::BindToUnit(unsigned int unit)
 {
   glActiveTexture(GL_TEXTURE0 + unit);
diff --git a/xbmc/guilib/TextureGL.h b/xbmc/guilib/TextureGL.h
index 58d219b62..0a979065e 100644
--- a/xbmc/guilib/TextureGL.h
+++ b/xbmc/guilib/TextureGL.h
@@ -24,6 +24,7 @@ public:
   void CreateTextureObject() override;
   void DestroyTextureObject() override;
   void LoadToGPU() override;
+  void SyncGPU() override;
   void BindToUnit(unsigned int unit) override;
 
 protected:
diff --git a/xbmc/settings/AdvancedSettings.cpp b/xbmc/settings/AdvancedSettings.cpp
index cc0624479..2d87b7efe 100644
--- a/xbmc/settings/AdvancedSettings.cpp
+++ b/xbmc/settings/AdvancedSettings.cpp
@@ -1220,6 +1220,10 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
     XMLUtils::GetInt(pElement, "algorithmdirtyregions",     m_guiAlgorithmDirtyRegions);
     XMLUtils::GetBoolean(pElement, "smartredraw", m_guiSmartRedraw);
     XMLUtils::GetBoolean(pElement, "transparentvideolayout", m_guiVideoLayoutTransparent);
+    XMLUtils::GetInt(pElement, "anisotropicfiltering", m_guiAnisotropicFiltering);
+    XMLUtils::GetBoolean(pElement, "fronttobackrendering", m_guiFrontToBackRendering);
+    XMLUtils::GetBoolean(pElement, "geometryclear", m_guiGeometryClear);
+    XMLUtils::GetBoolean(pElement, "asynctextureupload", m_guiAsyncTextureUpload);
   }
 
   std::string seekSteps;
diff --git a/xbmc/settings/AdvancedSettings.h b/xbmc/settings/AdvancedSettings.h
index 5abbed248..95e91cf93 100644
--- a/xbmc/settings/AdvancedSettings.h
+++ b/xbmc/settings/AdvancedSettings.h
@@ -333,6 +333,10 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     int  m_guiAlgorithmDirtyRegions;
     bool m_guiSmartRedraw;
     bool m_guiVideoLayoutTransparent{false};
+    int32_t m_guiAnisotropicFiltering{0};
+    bool m_guiFrontToBackRendering{false};
+    bool m_guiGeometryClear{true};
+    bool m_guiAsyncTextureUpload{false};
     unsigned int m_addonPackageFolderSize;
 
     bool m_jsonOutputCompact;
diff --git a/xbmc/utils/EGLUtils.cpp b/xbmc/utils/EGLUtils.cpp
index 85f889dc9..40338121d 100644
--- a/xbmc/utils/EGLUtils.cpp
+++ b/xbmc/utils/EGLUtils.cpp
@@ -428,6 +428,12 @@ bool CEGLContextUtils::CreateContext(CEGLAttributesVec contextAttribs)
     return false;
   }
 
+  m_eglUploadContext =
+      eglCreateContext(m_eglDisplay, m_eglConfig, m_eglContext, contextAttribs.Get());
+
+  if (m_eglUploadContext == EGL_NO_CONTEXT)
+    CLog::Log(LOGWARNING, "Failed to create EGL upload context");
+
   return true;
 }
 
@@ -565,6 +571,12 @@ void CEGLContextUtils::DestroyContext()
     eglDestroyContext(m_eglDisplay, m_eglContext);
     m_eglContext = EGL_NO_CONTEXT;
   }
+
+  if (m_eglUploadContext)
+  {
+    eglDestroyContext(m_eglDisplay, m_eglUploadContext);
+    m_eglUploadContext = EGL_NO_CONTEXT;
+  }
 }
 
 void CEGLContextUtils::DestroySurface()
@@ -597,3 +609,49 @@ bool CEGLContextUtils::TrySwapBuffers()
 
   return (eglSwapBuffers(m_eglDisplay, m_eglSurface) == EGL_TRUE);
 }
+
+bool CEGLContextUtils::BindTextureUploadContext()
+{
+  if (m_eglDisplay == EGL_NO_DISPLAY || m_eglUploadContext == EGL_NO_CONTEXT)
+  {
+    CLog::LogF(LOGERROR, "No texture upload context found.");
+    return false;
+  }
+
+  m_textureUploadLock.lock();
+
+  if (!eglMakeCurrent(m_eglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, m_eglUploadContext))
+  {
+    m_textureUploadLock.unlock();
+    CLog::LogF(LOGERROR, "Couldn't bind texture upload context.");
+    return false;
+  }
+
+  return true;
+}
+
+bool CEGLContextUtils::UnbindTextureUploadContext()
+{
+  if (m_eglDisplay == EGL_NO_DISPLAY || m_eglUploadContext == EGL_NO_CONTEXT)
+  {
+    CLog::LogF(LOGERROR, "No texture upload context found.");
+    m_textureUploadLock.unlock();
+    return false;
+  }
+
+  if (!eglMakeCurrent(m_eglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT))
+  {
+    CLog::LogF(LOGERROR, "Couldn't release texture upload context");
+    m_textureUploadLock.unlock();
+    return false;
+  }
+
+  m_textureUploadLock.unlock();
+
+  return true;
+}
+
+bool CEGLContextUtils::HasContext()
+{
+  return eglGetCurrentContext() != EGL_NO_CONTEXT;
+}
diff --git a/xbmc/utils/EGLUtils.h b/xbmc/utils/EGLUtils.h
index 2db162829..f611bcbd0 100644
--- a/xbmc/utils/EGLUtils.h
+++ b/xbmc/utils/EGLUtils.h
@@ -8,6 +8,8 @@
 
 #pragma once
 
+#include "threads/CriticalSection.h"
+
 #include <array>
 #include <set>
 #include <stdexcept>
@@ -219,6 +221,10 @@ public:
     return m_eglConfig;
   }
 
+  bool BindTextureUploadContext();
+  bool UnbindTextureUploadContext();
+  bool HasContext();
+
 private:
   void SurfaceAttrib();
 
@@ -229,4 +235,6 @@ private:
   EGLSurface m_eglSurface{EGL_NO_SURFACE};
   EGLContext m_eglContext{EGL_NO_CONTEXT};
   EGLConfig m_eglConfig{}, m_eglHDRConfig{};
+  EGLContext m_eglUploadContext{EGL_NO_CONTEXT};
+  mutable CCriticalSection m_textureUploadLock;
 };
diff --git a/xbmc/windowing/WinSystem.h b/xbmc/windowing/WinSystem.h
index 88218db9e..0724fca40 100644
--- a/xbmc/windowing/WinSystem.h
+++ b/xbmc/windowing/WinSystem.h
@@ -260,6 +260,24 @@ public:
    */
   std::pair<bool, int> GetDitherSettings();
 
+  /*!
+   * \brief Binds a shared context to the current thread, in order to upload textures asynchronously.
+   * \return Return true if a texture upload context exists and the binding succeeds.
+   */
+  virtual bool BindTextureUploadContext() { return false; }
+
+  /*!
+   * \brief Unbinds the shared context.
+   * \return Return true if the texture upload context has been unbound.
+   */
+  virtual bool UnbindTextureUploadContext() { return false; }
+
+  /*!
+   * \brief Checks if a graphics context is already bound to the current thread.
+   * \return Return true if so.
+   */
+  virtual bool HasContext() { return false; }
+
 protected:
   void UpdateDesktopResolution(RESOLUTION_INFO& newRes, const std::string &output, int width, int height, float refreshRate, uint32_t dwFlags);
   void UpdateDesktopResolution(RESOLUTION_INFO& newRes,
diff --git a/xbmc/windowing/X11/GLContextEGL.cpp b/xbmc/windowing/X11/GLContextEGL.cpp
index 7999e76b9..6a0352f9e 100644
--- a/xbmc/windowing/X11/GLContextEGL.cpp
+++ b/xbmc/windowing/X11/GLContextEGL.cpp
@@ -14,6 +14,7 @@
 #include "GLContextEGL.h"
 
 #include "ServiceBroker.h"
+#include "commons/ilog.h"
 #include "settings/AdvancedSettings.h"
 #include "settings/SettingsComponent.h"
 #include "utils/log.h"
@@ -21,12 +22,13 @@
 #include <clocale>
 #include <mutex>
 
-#include <EGL/eglext.h>
 #include <unistd.h>
 
 #include "PlatformDefs.h"
 #include "system_gl.h"
 
+#include <EGL/eglext.h>
+
 #define EGL_NO_CONFIG (EGLConfig)0
 
 CGLContextEGL::CGLContextEGL(Display* dpy, EGLint renderingApi)
@@ -160,6 +162,8 @@ bool CGLContextEGL::Refresh(bool force, int screen, Window glWindow, bool &newCo
       EGL_NONE
   };
   m_eglContext = eglCreateContext(m_eglDisplay, m_eglConfig, EGL_NO_CONTEXT, contextAttributes);
+  m_eglUploadContext = eglCreateContext(m_eglDisplay, m_eglConfig, m_eglContext, contextAttributes);
+
   if (m_eglContext == EGL_NO_CONTEXT)
   {
     EGLint contextAttributes[] =
@@ -168,6 +172,8 @@ bool CGLContextEGL::Refresh(bool force, int screen, Window glWindow, bool &newCo
       EGL_NONE
     };
     m_eglContext = eglCreateContext(m_eglDisplay, m_eglConfig, EGL_NO_CONTEXT, contextAttributes);
+    m_eglUploadContext =
+        eglCreateContext(m_eglDisplay, m_eglConfig, m_eglContext, contextAttributes);
 
     if (m_eglContext == EGL_NO_CONTEXT)
     {
@@ -190,6 +196,9 @@ bool CGLContextEGL::Refresh(bool force, int screen, Window glWindow, bool &newCo
 
   m_eglGetSyncValuesCHROMIUM = (PFNEGLGETSYNCVALUESCHROMIUMPROC)eglGetProcAddress("eglGetSyncValuesCHROMIUM");
 
+  if (m_eglUploadContext == EGL_NO_CONTEXT)
+    CLog::Log(LOGWARNING, "failed to create EGL upload context");
+
   m_usePB = false;
   return true;
 }
@@ -286,6 +295,11 @@ bool CGLContextEGL::CreatePB()
     return false;
   }
 
+  m_eglUploadContext = eglCreateContext(m_eglDisplay, m_eglConfig, m_eglContext, contextAttributes);
+
+  if (m_eglUploadContext == EGL_NO_CONTEXT)
+    CLog::Log(LOGWARNING, "Failed to create EGL upload context");
+
   m_usePB = true;
   return true;
 }
@@ -300,6 +314,12 @@ void CGLContextEGL::Destroy()
     m_eglContext = EGL_NO_CONTEXT;
   }
 
+  if (m_eglUploadContext)
+  {
+    eglDestroyContext(m_eglDisplay, m_eglUploadContext);
+    m_eglUploadContext = EGL_NO_CONTEXT;
+  }
+
   if (m_eglSurface)
   {
     eglDestroySurface(m_eglDisplay, m_eglSurface);
@@ -534,6 +554,52 @@ uint64_t CGLContextEGL::GetVblankTiming(uint64_t &msc, uint64_t &interval)
   return ret;
 }
 
+bool CGLContextEGL::BindTextureUploadContext()
+{
+  if (m_eglDisplay == EGL_NO_DISPLAY || m_eglUploadContext == EGL_NO_CONTEXT)
+  {
+    CLog::LogF(LOGERROR, "No texture upload context found.");
+    return false;
+  }
+
+  m_textureUploadLock.lock();
+
+  if (!eglMakeCurrent(m_eglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, m_eglUploadContext))
+  {
+    m_textureUploadLock.unlock();
+    CLog::LogF(LOGERROR, "Couldn't bind texture upload context.");
+    return false;
+  }
+
+  return true;
+}
+
+bool CGLContextEGL::UnbindTextureUploadContext()
+{
+  if (m_eglDisplay == EGL_NO_DISPLAY || m_eglUploadContext == EGL_NO_CONTEXT)
+  {
+    CLog::LogF(LOGERROR, "No texture upload context found.");
+    m_textureUploadLock.unlock();
+    return false;
+  }
+
+  if (!eglMakeCurrent(m_eglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT))
+  {
+    CLog::LogF(LOGERROR, "Couldn't release texture upload context");
+    m_textureUploadLock.unlock();
+    return false;
+  }
+
+  m_textureUploadLock.unlock();
+
+  return true;
+}
+
+bool CGLContextEGL::HasContext()
+{
+  return eglGetCurrentContext() != EGL_NO_CONTEXT;
+}
+
 void CGLContextEGL::QueryExtensions()
 {
   std::string extensions = eglQueryString(m_eglDisplay, EGL_EXTENSIONS);
diff --git a/xbmc/windowing/X11/GLContextEGL.h b/xbmc/windowing/X11/GLContextEGL.h
index 441787b67..7c38f8125 100644
--- a/xbmc/windowing/X11/GLContextEGL.h
+++ b/xbmc/windowing/X11/GLContextEGL.h
@@ -9,11 +9,12 @@
 #pragma once
 
 #include "GLContext.h"
-#include "system_egl.h"
 #include "threads/CriticalSection.h"
 
 #include <cstdint>
 
+#include "system_egl.h"
+
 #include <EGL/eglext.h>
 #ifdef HAVE_EGLEXTANGLE
 #include <EGL/eglext_angle.h>
@@ -36,6 +37,10 @@ public:
   void QueryExtensions() override;
   uint64_t GetVblankTiming(uint64_t &msc, uint64_t &interval) override;
 
+  bool BindTextureUploadContext();
+  bool UnbindTextureUploadContext();
+  bool HasContext();
+
   EGLint m_renderingApi;
   EGLDisplay m_eglDisplay = EGL_NO_DISPLAY;
   EGLSurface m_eglSurface = EGL_NO_SURFACE;
@@ -60,4 +65,7 @@ protected:
   CCriticalSection m_syncLock;
 
   bool m_usePB = false;
+
+  EGLContext m_eglUploadContext = EGL_NO_CONTEXT;
+  mutable CCriticalSection m_textureUploadLock;
 };
diff --git a/xbmc/windowing/X11/WinSystemX11GLContext.cpp b/xbmc/windowing/X11/WinSystemX11GLContext.cpp
index 44ddb8985..8b3fda053 100644
--- a/xbmc/windowing/X11/WinSystemX11GLContext.cpp
+++ b/xbmc/windowing/X11/WinSystemX11GLContext.cpp
@@ -112,6 +112,30 @@ EGLConfig CWinSystemX11GLContext::GetEGLConfig() const
   return static_cast<CGLContextEGL*>(m_pGLContext)->m_eglConfig;
 }
 
+bool CWinSystemX11GLContext::BindTextureUploadContext()
+{
+  if (m_pGLContext)
+    return static_cast<CGLContextEGL*>(m_pGLContext)->BindTextureUploadContext();
+  else
+    return false;
+}
+
+bool CWinSystemX11GLContext::UnbindTextureUploadContext()
+{
+  if (m_pGLContext)
+    return static_cast<CGLContextEGL*>(m_pGLContext)->UnbindTextureUploadContext();
+  else
+    return false;
+}
+
+bool CWinSystemX11GLContext::HasContext()
+{
+  if (m_pGLContext)
+    return static_cast<CGLContextEGL*>(m_pGLContext)->HasContext();
+  else
+    return false;
+}
+
 bool CWinSystemX11GLContext::SetWindow(int width, int height, bool fullscreen, const std::string &output, int *winstate)
 {
   int newwin = 0;
diff --git a/xbmc/windowing/X11/WinSystemX11GLContext.h b/xbmc/windowing/X11/WinSystemX11GLContext.h
index eda6772a4..2c11b9abe 100644
--- a/xbmc/windowing/X11/WinSystemX11GLContext.h
+++ b/xbmc/windowing/X11/WinSystemX11GLContext.h
@@ -58,6 +58,10 @@ public:
   EGLContext GetEGLContext() const;
   EGLConfig GetEGLConfig() const;
 
+  bool BindTextureUploadContext() override;
+  bool UnbindTextureUploadContext() override;
+  bool HasContext() override;
+
 protected:
   bool SetWindow(int width, int height, bool fullscreen, const std::string &output, int *winstate = NULL) override;
   void PresentRenderImpl(bool rendered) override;
diff --git a/xbmc/windowing/X11/WinSystemX11GLESContext.cpp b/xbmc/windowing/X11/WinSystemX11GLESContext.cpp
index e2367d3e6..436f75a09 100644
--- a/xbmc/windowing/X11/WinSystemX11GLESContext.cpp
+++ b/xbmc/windowing/X11/WinSystemX11GLESContext.cpp
@@ -92,6 +92,30 @@ EGLConfig CWinSystemX11GLESContext::GetEGLConfig() const
   return m_pGLContext->m_eglConfig;
 }
 
+bool CWinSystemX11GLESContext::BindTextureUploadContext()
+{
+  if (m_pGLContext)
+    return static_cast<CGLContextEGL*>(m_pGLContext)->BindTextureUploadContext();
+  else
+    return false;
+}
+
+bool CWinSystemX11GLESContext::UnbindTextureUploadContext()
+{
+  if (m_pGLContext)
+    return static_cast<CGLContextEGL*>(m_pGLContext)->UnbindTextureUploadContext();
+  else
+    return false;
+}
+
+bool CWinSystemX11GLESContext::HasContext()
+{
+  if (m_pGLContext)
+    return static_cast<CGLContextEGL*>(m_pGLContext)->HasContext();
+  else
+    return false;
+}
+
 bool CWinSystemX11GLESContext::SetWindow(int width, int height, bool fullscreen, const std::string& output, int* winstate)
 {
   int newwin = 0;
diff --git a/xbmc/windowing/X11/WinSystemX11GLESContext.h b/xbmc/windowing/X11/WinSystemX11GLESContext.h
index cdb1cc4f8..1f066247b 100644
--- a/xbmc/windowing/X11/WinSystemX11GLESContext.h
+++ b/xbmc/windowing/X11/WinSystemX11GLESContext.h
@@ -46,6 +46,10 @@ public:
   EGLContext GetEGLContext() const;
   EGLConfig GetEGLConfig() const;
 
+  bool BindTextureUploadContext() override;
+  bool UnbindTextureUploadContext() override;
+  bool HasContext() override;
+
 protected:
   bool SetWindow(int width, int height, bool fullscreen, const std::string& output, int* winstate = nullptr) override;
   void PresentRenderImpl(bool rendered) override;
diff --git a/xbmc/windowing/gbm/WinSystemGbmEGLContext.cpp b/xbmc/windowing/gbm/WinSystemGbmEGLContext.cpp
index fac914d5e..b7e02d58b 100644
--- a/xbmc/windowing/gbm/WinSystemGbmEGLContext.cpp
+++ b/xbmc/windowing/gbm/WinSystemGbmEGLContext.cpp
@@ -150,3 +150,18 @@ void CWinSystemGbmEGLContext::delete_CVaapiProxy::operator()(CVaapiProxy *p) con
 {
   VaapiProxyDelete(p);
 }
+
+bool CWinSystemGbmEGLContext::BindTextureUploadContext()
+{
+  return m_eglContext.BindTextureUploadContext();
+}
+
+bool CWinSystemGbmEGLContext::UnbindTextureUploadContext()
+{
+  return m_eglContext.UnbindTextureUploadContext();
+}
+
+bool CWinSystemGbmEGLContext::HasContext()
+{
+  return m_eglContext.HasContext();
+}
diff --git a/xbmc/windowing/gbm/WinSystemGbmEGLContext.h b/xbmc/windowing/gbm/WinSystemGbmEGLContext.h
index fbd52354e..e8b0a76d4 100644
--- a/xbmc/windowing/gbm/WinSystemGbmEGLContext.h
+++ b/xbmc/windowing/gbm/WinSystemGbmEGLContext.h
@@ -35,6 +35,10 @@ public:
                        RESOLUTION_INFO& res) override;
   bool DestroyWindow() override;
 
+  bool BindTextureUploadContext() override;
+  bool UnbindTextureUploadContext() override;
+  bool HasContext() override;
+
 protected:
   CWinSystemGbmEGLContext(EGLenum platform, std::string const& platformExtension)
     : CWinSystemEGL{platform, platformExtension}
diff --git a/xbmc/windowing/wayland/WinSystemWaylandEGLContext.cpp b/xbmc/windowing/wayland/WinSystemWaylandEGLContext.cpp
index 6750c547d..f972ad549 100644
--- a/xbmc/windowing/wayland/WinSystemWaylandEGLContext.cpp
+++ b/xbmc/windowing/wayland/WinSystemWaylandEGLContext.cpp
@@ -102,6 +102,21 @@ bool CWinSystemWaylandEGLContext::DestroyWindowSystem()
   return CWinSystemWaylandImpl::DestroyWindowSystem();
 }
 
+bool CWinSystemWaylandEGLContext::BindTextureUploadContext()
+{
+  return m_eglContext.BindTextureUploadContext();
+}
+
+bool CWinSystemWaylandEGLContext::UnbindTextureUploadContext()
+{
+  return m_eglContext.UnbindTextureUploadContext();
+}
+
+bool CWinSystemWaylandEGLContext::HasContext()
+{
+  return m_eglContext.HasContext();
+}
+
 CSizeInt CWinSystemWaylandEGLContext::GetNativeWindowAttachedSize()
 {
   int width, height;
diff --git a/xbmc/windowing/wayland/WinSystemWaylandEGLContext.h b/xbmc/windowing/wayland/WinSystemWaylandEGLContext.h
index 68511f4ff..a6e0a1440 100644
--- a/xbmc/windowing/wayland/WinSystemWaylandEGLContext.h
+++ b/xbmc/windowing/wayland/WinSystemWaylandEGLContext.h
@@ -43,6 +43,10 @@ public:
   bool DestroyWindow() override;
   bool DestroyWindowSystem() override;
 
+  bool BindTextureUploadContext() override;
+  bool UnbindTextureUploadContext() override;
+  bool HasContext() override;
+
 protected:
   /**
    * Inheriting classes should override InitWindowSystem() without parameters
